
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>phydmslib package &#8212; phydms 2.3.8 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Acknowledgments and citations" href="acknowledgments.html" />
    <link rel="prev" title="phydmslib API" href="api.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="phydmslib-package">
<h1>phydmslib package<a class="headerlink" href="#phydmslib-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-phydmslib.constants">
<span id="phydmslib-constants-module"></span><h2>phydmslib.constants module<a class="headerlink" href="#module-phydmslib.constants" title="Permalink to this headline">¶</a></h2>
<p>Defines constants and nucleotide / amino acid / codon indices.</p>
<dl class="simple">
<dt>Constants defined:</dt><dd><dl class="simple">
<dt><cite>ALMOST_ZERO</cite> (float)</dt><dd><p>a small number used to check for quantities close to zero</p>
</dd>
<dt><cite>N_NT</cite> (int)</dt><dd><p>number of nucleotides</p>
</dd>
<dt><cite>N_AA</cite> (int)</dt><dd><p>number of amino acids</p>
</dd>
<dt><cite>N_CODON</cite> (int)</dt><dd><p>number of codons</p>
</dd>
<dt><cite>N_STOP</cite> (int)</dt><dd><p>number of stop codons</p>
</dd>
<dt><cite>NT_TO_INDEX</cite> (dict)</dt><dd><p>mapping of one-letter nucleotides to integer indices</p>
</dd>
<dt><cite>INDEX_TO_NT</cite> (dict)</dt><dd><p>inverse of <cite>NT_TO_INDEX</cite></p>
</dd>
<dt><cite>AA_TO_INDEX</cite> (dict)</dt><dd><p>mapping of one-letter amino acids to integer indices</p>
</dd>
<dt><cite>INDEX_TO_AA</cite> (dict)</dt><dd><p>inverse of <cite>AA_TO_INDEX</cite></p>
</dd>
<dt><cite>CODON_TO_INDEX</cite> (dict):</dt><dd><p>mapping of non-stop codons to integer indices</p>
</dd>
<dt><cite>INDEX_TO_CODON</cite> (dict)</dt><dd><p>inverse of <cite>CODON_TO_INDEX</cite></p>
</dd>
<dt><cite>CODON_TO_AA</cite> (<cite>numpy.ndarray</cite> of int, length <cite>N_CODON</cite>)</dt><dd><p>Element <cite>x</cite> gives index for amino acid encoded by <cite>x</cite>.</p>
</dd>
<dt><cite>PURINES</cite> (<cite>frozenset</cite>)</dt><dd><p>Set of one-letter nucleotides for purines</p>
</dd>
<dt><cite>PYRIMIDINES</cite> (<cite>frozenset</cite>)</dt><dd><p>Set of one-letter nucleotides for pyrimidines</p>
</dd>
<dt><cite>CODON_TRANSITION</cite> (<cite>numpy.ndarray</cite> of bools, shape <cite>(N_CODON, NCODON)</cite>)</dt><dd><p>Element <cite>[x][y]</cite> is <cite>True</cite> iff <cite>x</cite> and <cite>y</cite> differ by 1 transition.</p>
</dd>
<dt><cite>CODON_SINGLEMUT</cite> (<cite>numpy.ndarray</cite> of bools, shape (<cite>N_CODON, N_CODON)</cite>)</dt><dd><p>Element <cite>[x][y]</cite> is <cite>True</cite> iff <cite>x</cite> and <cite>y</cite> differ by 1 nt mutation.</p>
</dd>
<dt><cite>CODON_NT_MUT</cite> (<cite>numpy.ndarray</cite> of bools, shape <cite>(N_NT, N_CODON, N_CODON)</cite>)</dt><dd><p>Element <cite>[w][x][y]</cite> is <cite>True</cite> iff <cite>x</cite> converts to <cite>y</cite> by single
nucleotide mutation <cite>w</cite>.</p>
</dd>
<dt><cite>CODON_NT</cite> (<cite>numpy.ndarray</cite> of bools, shape <cite>(3, N_NT, N_CODON)</cite>)</dt><dd><p>Element <cite>[j][w][x]</cite> is <cite>True</cite> iff nt <cite>j</cite> of codon <cite>x</cite> is <cite>w</cite>.</p>
</dd>
<dt><cite>CODON_NT_INDEX</cite> (<cite>numpy.ndarray</cite> of int, shape <cite>(3, N_CODON)</cite>)</dt><dd><p>Element <cite>[j][x]</cite> gives index of nt at position <cite>j</cite> of codon <cite>x</cite>.</p>
</dd>
<dt><cite>CODON_NONSYN</cite> (<cite>numpy.ndarray</cite> of bools, shape <cite>(N_CODON, N_CODON)</cite>)</dt><dd><p>Element <cite>[x][y]</cite> is <cite>True</cite> iff mutating  <cite>x</cite> to <cite>y</cite> is nonsynonymous</p>
</dd>
<dt><cite>CODON_NT_COUNT</cite> (<cite>numpy.ndarray</cite> of int, shape <cite>(N_NT, N_CODON)</cite>)</dt><dd><p>Element <cite>[w][x]</cite> gives the number of occurrences of nucleotide
<cite>w</cite> in codon <cite>x</cite>.</p>
</dd>
<dt><cite>STOP_CODON_TO_NT_INDICES</cite> (<cite>numpy.ndarray</cite> of float, shape <cite>(N_STOP, 3, N_NT)</cite>)</dt><dd><p>Element <cite>[x][p][w]</cite> is 1.0 if codon position <cite>p</cite> is nucleotide <cite>w</cite>
in stop codon <cite>x</cite> and 0.0 otherwise.</p>
</dd>
<dt><cite>STOP_POSITIONS</cite> (<cite>numpy.ndarray</cite> of float, shape <cite>(3, N_NT)</cite>)</dt><dd><p>Element <cite>[p][w]</cite> is -1.0 if any stop codon has nucleotide <cite>w</cite> in
codon position <cite>p</cite> and 1.0 otherwise</p>
</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="module-phydmslib.file_io">
<span id="phydmslib-file-io-module"></span><h2>phydmslib.file_io module<a class="headerlink" href="#module-phydmslib.file_io" title="Permalink to this headline">¶</a></h2>
<p>Module for input / output from files.</p>
<dl class="function">
<dt id="phydmslib.file_io.ReadCodonAlignment">
<code class="sig-prename descclassname">phydmslib.file_io.</code><code class="sig-name descname">ReadCodonAlignment</code><span class="sig-paren">(</span><em class="sig-param">fastafile</em>, <em class="sig-param">checknewickvalid</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.file_io.ReadCodonAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads codon alignment from file.</p>
<p><em>fastafile</em> is the name of an existing FASTA file.</p>
<p><em>checknewickvalid</em> : if <em>True</em>, we require that names are unique and do
<strong>not</strong> contain spaces, commas, colons, semicolons, parentheses, square
brackets, or single or double quotation marks.
If any of these disallowed characters are present, raises an Exception.</p>
<p>Reads the alignment from the <em>fastafile</em> and returns the aligned
sequences as a list of 2-tuple of strings <em>(header, sequence)</em>
where <em>sequence</em> is upper case.</p>
<p>If the terminal codon is a stop codon for <strong>all</strong> sequences, then
this terminal codon is trimmed. Raises an exception if the sequences
are not aligned codon sequences that are free of stop codons (with
the exception of a shared terminal stop) and free of ambiguous nucleotides.</p>
<p>Read aligned sequences in this example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;ATGGAA&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;ATGAAA&#39;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">u</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&gt;</span><span class="si">{0}</span><span class="se">\n</span><span class="si">{1}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">*</span><span class="n">tup</span><span class="p">)</span> <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">ReadCodonAlignment</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">==</span> <span class="n">a</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Trim stop codons from all sequences in this example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;ATGTAA&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;ATGTGA&#39;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">u</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&gt;</span><span class="si">{0}</span><span class="se">\n</span><span class="si">{1}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">*</span><span class="n">tup</span><span class="p">)</span> <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">ReadCodonAlignment</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[(</span><span class="n">head</span><span class="p">,</span> <span class="n">seq</span><span class="p">[</span> <span class="p">:</span> <span class="o">-</span><span class="mi">3</span><span class="p">])</span> <span class="k">for</span> <span class="p">(</span><span class="n">head</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">]</span> <span class="o">==</span> <span class="n">a</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Read sequences with gap:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;ATG---&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;ATGAGA&#39;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">u</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&gt;</span><span class="si">{0}</span><span class="se">\n</span><span class="si">{1}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">*</span><span class="n">tup</span><span class="p">)</span> <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">ReadCodonAlignment</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[(</span><span class="n">head</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">head</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">]</span> <span class="o">==</span> <span class="n">a</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Premature stop codon gives error:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seqs</span> <span class="o">=</span> <span class="p">[(</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;TGAATG&#39;</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;ATGAGA&#39;</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">io</span><span class="o">.</span><span class="n">StringIO</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">u</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;&gt;</span><span class="si">{0}</span><span class="se">\n</span><span class="si">{1}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">*</span><span class="n">tup</span><span class="p">)</span> <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">seqs</span><span class="p">]))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">ReadCodonAlignment</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span> 
<span class="gt">Traceback (most recent call last):</span>
<span class="gr">ValueError</span>:
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="phydmslib.file_io.Versions">
<code class="sig-prename descclassname">phydmslib.file_io.</code><code class="sig-name descname">Versions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.file_io.Versions" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a string with version information.</p>
<p>You would call this function if you want a string giving detailed information
on the version of <code class="docutils literal notranslate"><span class="pre">phydms</span></code> and the associated packages that it uses.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.file_io.readDivPressure">
<code class="sig-prename descclassname">phydmslib.file_io.</code><code class="sig-name descname">readDivPressure</code><span class="sig-paren">(</span><em class="sig-param">fileName</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.file_io.readDivPressure" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads in diversifying pressures from some file.</p>
<p>Scale diversifying pressure values so absolute value of the max value is 1,
unless all values are zero.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>fileName</cite> (string or readable file-like object)</dt><dd><p>File holding diversifying pressure values. Can be
comma-, space-, or tab-separated file. The first column
is the site (consecutively numbered, sites starting
with one) and the second column is the diversifying pressure values.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>divPressure</cite> (dict keyed by ints)</dt><dd><p><cite>divPressure[r][v]</cite> is the diversifying pressure value of site <cite>r</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="phydmslib.file_io.readPrefs">
<code class="sig-prename descclassname">phydmslib.file_io.</code><code class="sig-name descname">readPrefs</code><span class="sig-paren">(</span><em class="sig-param">prefsfile</em>, <em class="sig-param">minpref=0</em>, <em class="sig-param">avgprefs=False</em>, <em class="sig-param">randprefs=False</em>, <em class="sig-param">seed=1</em>, <em class="sig-param">sites_as_strings=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.file_io.readPrefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Read preferences from file with some error checking.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefsfile</cite> (string or readable file-like object)</dt><dd><p>File holding amino-acid preferences. Can be
comma-, space-, or tab-separated file with column
headers of <cite>site</cite> and then all one-letter amino-acid
codes, or can be in the more complex format written
<a class="reference external" href="http://jbloomlab.github.io/dms_tools/">dms_tools v1</a>.
Must be prefs for consecutively numbered sites starting at 1.
Stop codon prefs can be present (stop codons are indicated by
<code class="docutils literal notranslate"><span class="pre">*</span></code>); if so they are removed and prefs re-normalized to sum to 1.</p>
</dd>
<dt><cite>minpref</cite> (float &gt;= 0)</dt><dd><p>Adjust all preferences to be &gt;= this number.</p>
</dd>
<dt><cite>avgprefs</cite>, <cite>randprefs</cite> (bool)</dt><dd><p>Mutually exclusive options specifying to average or
randomize prefs across sites.</p>
</dd>
<dt><cite>seed</cite> (int)</dt><dd><p>Seed used to sort random number generator for <cite>randprefs</cite>.</p>
</dd>
<dt><cite>sites_as_strings</cite> (bool)</dt><dd><p>By default, the site numers are coerced to integers.
If this option is <cite>True</cite>, then they are kept as strings.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>prefs</cite> (dict)</dt><dd><p><cite>prefs[r][a]</cite> is the preference of site <cite>r</cite> for amino-acid <cite>a</cite>.
<cite>r</cite> is an <cite>int</cite> unless <cite>sites_as_strings=True</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="phydmslib.file_io.readPrefs_dms_tools_format">
<code class="sig-prename descclassname">phydmslib.file_io.</code><code class="sig-name descname">readPrefs_dms_tools_format</code><span class="sig-paren">(</span><em class="sig-param">f</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.file_io.readPrefs_dms_tools_format" title="Permalink to this definition">¶</a></dt>
<dd><p>Reads the amino-acid preferences written by <a class="reference external" href="http://jbloomlab.github.io/dms_tools/">dms_tools v1</a>.</p>
<p>This is an exact copy of the same code from
<cite>dms_tools.file_io.ReadPreferences</cite>. It is copied because
<a class="reference external" href="http://jbloomlab.github.io/dms_tools/">dms_tools v1</a> is currently
only compatible with <cite>python2</cite>, and we needed something that also works
with <cite>python3</cite>.</p>
<p><em>f</em> is the name of an existing file or a readable file-like object.
It should be in the format written by
<a class="reference external" href="http://jbloomlab.github.io/dms_tools/">dms_tools v1</a>.</p>
<p>The return value is the tuple: <em>(sites, wts, pi_means, pi_95credint, h)</em>
where <em>sites</em>, <em>wts</em>, <em>pi_means</em>, and <em>pi_95credint</em> will all
have the same values used to write the file with <em>WritePreferences</em>,
and <em>h</em> is a dictionary with <em>h[r]</em> giving the site entropy (log base
2) for each <em>r</em> in <em>sites</em>.</p>
</dd></dl>

</div>
<div class="section" id="module-phydmslib.models">
<span id="phydmslib-models-module"></span><h2>phydmslib.models module<a class="headerlink" href="#module-phydmslib.models" title="Permalink to this headline">¶</a></h2>
<p>Substitution models.</p>
<p>All models defined here are subclasses of abstract base class <cite>Model</cite>.</p>
<p>Nucleotides, amino acids, and codons are indexed by integers 0, 1, …
using the indexing schemes defined in <cite>phydmslib.constants</cite>.</p>
<dl class="function">
<dt id="phydmslib.models.DiscreteGamma">
<code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">DiscreteGamma</code><span class="sig-paren">(</span><em class="sig-param">alpha</em>, <em class="sig-param">beta</em>, <em class="sig-param">ncats</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.DiscreteGamma" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns category means for discretized gamma distribution.</p>
<p>The distribution is evenly divided into categories, and the
mean of each category is returned.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>alpha</cite> (<cite>float</cite> &gt; 0)</dt><dd><p>Shape parameter of gamma distribution.</p>
</dd>
<dt><cite>beta</cite> (<cite>float</cite> &gt; 0)</dt><dd><p>Inverse scale parameter of gamma distribution.</p>
</dd>
<dt><cite>ncats</cite> (<cite>int</cite> &gt; 1)</dt><dd><p>Number of categories in discretized gamma distribution.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>catmeans</cite> (<cite>scipy.ndarray</cite> of <cite>float</cite>, shape <cite>(ncats,)</cite>)</dt><dd><p><cite>catmeans[k]</cite> is the mean of category <cite>k</cite> where
<cite>0 &lt;= k &lt; ncats</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<p>Check that we get values in Figure 1 of Yang, J Mol Evol, 39:306-314
&gt;&gt;&gt; catmeans = DiscreteGamma(0.5, 0.5, 4)
&gt;&gt;&gt; numpy.allclose(catmeans, numpy.array([0.0334, 0.2519, 0.8203, 2.8944]), atol=1e-4)
True</p>
<p>Make sure we get expected mean of alpha / beta
&gt;&gt;&gt; alpha = 0.6
&gt;&gt;&gt; beta = 0.8
&gt;&gt;&gt; ncats = 6
&gt;&gt;&gt; catmeans = DiscreteGamma(alpha, beta, ncats)
&gt;&gt;&gt; numpy.allclose(catmeans.sum() / ncats, alpha / beta)
True</p>
</dd></dl>

<dl class="class">
<dt id="phydmslib.models.DistributionModel">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">DistributionModel</code><a class="headerlink" href="#phydmslib.models.DistributionModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.Model" title="phydmslib.models.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.Model</span></code></a></p>
<p>Substitution model with a parameter drawn from distribution.</p>
<p>This abstract base class defines required methods / attributes
of substitution models with <em>one</em> parameter (<cite>distributedparam</cite>)
drawn from a distribution. An example is that <cite>omega</cite> might be
drawn from a gamma distribution.</p>
<dl class="method">
<dt id="phydmslib.models.DistributionModel.M">
<em class="property">abstract </em><code class="sig-name descname">M</code><span class="sig-paren">(</span><em class="sig-param">k</em>, <em class="sig-param">t</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.DistributionModel.M" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix exponential <cite>M(mu * t) = exp(mu * t * P)</cite> for category <cite>k</cite>.</p>
<p>Similar to definition for <cite>Model</cite> abstract base class except
also specifies for which category <cite>k</cite> we are returning the
matrix exponential. <cite>k</cite> is an <cite>int</cite> with <cite>0 &lt;= k &lt; ncats</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.basemodel">
<em class="property">abstract property </em><code class="sig-name descname">basemodel</code><a class="headerlink" href="#phydmslib.models.DistributionModel.basemodel" title="Permalink to this definition">¶</a></dt>
<dd><p>Base model over which this model is distributed.</p>
<p>A <cite>DistributionModel</cite> is a base model with some property
drawn from a distribution. This gives such a base model.
Parameters of this base model that are <strong>not</strong> distributed
among sites will be current, but the distributed parameter
will not.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.catweights">
<em class="property">abstract property </em><code class="sig-name descname">catweights</code><a class="headerlink" href="#phydmslib.models.DistributionModel.catweights" title="Permalink to this definition">¶</a></dt>
<dd><p>Weight assigned to each category.</p>
<p>Is a <cite>numpy.ndarray</cite> of <cite>float</cite> of shape <cite>(ncats,)</cite> where each
entry is &gt; 0 and the entries sum to one. <cite>catweights[k]</cite> is
the weight assigned to category <cite>k</cite> where <cite>0 &lt;= k &lt; ncats</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.dM">
<em class="property">abstract </em><code class="sig-name descname">dM</code><span class="sig-paren">(</span><em class="sig-param">k</em>, <em class="sig-param">t</em>, <em class="sig-param">param</em>, <em class="sig-param">Mkt</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.DistributionModel.dM" title="Permalink to this definition">¶</a></dt>
<dd><p>Derivative of <cite>M(k, t)</cite> with respect to <cite>param</cite>.</p>
<p>Similar to definiton for <cite>Model</cite> abstract base class
except also specifies for which category <cite>k</cite> we are
returning the derivative. <cite>k</cite> is an <cite>int</cite> with
<cite>0 &lt;= k &lt; ncats</cite>.</p>
<p>Also, <cite>param</cite> can be any string in <cite>freeparams</cite> <strong>except</strong>
those in <cite>distributionparams</cite>, and can also be <cite>distributedparam</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.d_distributionparams">
<em class="property">abstract property </em><code class="sig-name descname">d_distributionparams</code><a class="headerlink" href="#phydmslib.models.DistributionModel.d_distributionparams" title="Permalink to this definition">¶</a></dt>
<dd><p>Derivatives of <cite>distributedparam</cite> with respect to <cite>distributionparams</cite>.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>d_distributionparams</cite> (<cite>dict</cite> keyed by params in <cite>distributionparams</cite>)</dt><dd><p><cite>d_distribution[param]</cite> is a <cite>numpy.ndarray</cite> of <cite>float</cite> of
shape <cite>(ncats,)</cite>, with <cite>d_distributionparam[param][k]</cite> giving
the derivative of the <cite>distributedparam</cite> for category <cite>k</cite>
with respect to <cite>param</cite> where <cite>param</cite> is in <cite>distributionparams</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.distributedparam">
<em class="property">abstract property </em><code class="sig-name descname">distributedparam</code><a class="headerlink" href="#phydmslib.models.DistributionModel.distributedparam" title="Permalink to this definition">¶</a></dt>
<dd><p>String giving name of parameter drawn from the distribution.</p>
<p>This parameter must be in <cite>freeparams</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.distributionparams">
<em class="property">abstract property </em><code class="sig-name descname">distributionparams</code><a class="headerlink" href="#phydmslib.models.DistributionModel.distributionparams" title="Permalink to this definition">¶</a></dt>
<dd><p>List of params giving shape of distribution for <cite>distributedparam</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.dstationarystate">
<em class="property">abstract </em><code class="sig-name descname">dstationarystate</code><span class="sig-paren">(</span><em class="sig-param">k</em>, <em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.DistributionModel.dstationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>Derivative of <cite>stationarystate</cite> with respect to <cite>param</cite> for category <cite>k</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>k</cite> (int)</dt><dd><p>Category for which we are computing stationary state.</p>
</dd>
<dt><cite>param</cite> (string)</dt><dd><p>A string in <cite>freeparams</cite> or <cite>distributedparam</cite></p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>dstationarystate</cite> (<cite>numpy.ndarray</cite> of floats or zero)</dt><dd><p>If <cite>param</cite> is a float, then <cite>dstationarystate[k][r][x]</cite>
is derivative of <cite>stationarystate[k][r][x]</cite> with respect
to <cite>param</cite> for category <cite>k</cite>. If <cite>param</cite> is an array, then
<cite>dstationarystate[k][i][r][x]</cite> is derivative of
<cite>stationarystate[k][r][x]</cite> with respect to <cite>param[i]</cite> for
category <cite>k</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.ncats">
<em class="property">abstract property </em><code class="sig-name descname">ncats</code><a class="headerlink" href="#phydmslib.models.DistributionModel.ncats" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of categories for distributed parameter.</p>
<p>Is an <cite>int</cite> &gt; 1.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.DistributionModel.stationarystate">
<em class="property">abstract </em><code class="sig-name descname">stationarystate</code><span class="sig-paren">(</span><em class="sig-param">k</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.DistributionModel.stationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>Stationary state of substitution model.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>k</cite> (int)</dt><dd><p>Category for which we are computing stationary state.</p>
</dd>
</dl>
</dd>
</dl>
<p>A <cite>numpy.ndarray</cite> of floats, shape <cite>(k, nsites, N_CODON)</cite>.
Element <cite>stationarystate[k][r][x]</cite> is the stationary
state probability of codon <cite>x</cite> at site <cite>r</cite> for category <cite>k</cite>.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.ExpCM">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">ExpCM</code><span class="sig-paren">(</span><em class="sig-param">prefs, kappa=2.0, omega=0.5, beta=1.0, mu=1.0, phi=array([0.25, 0.25, 0.25, 0.25]), freeparams=['kappa', 'omega', 'beta', 'mu', 'eta']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.Model" title="phydmslib.models.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.Model</span></code></a></p>
<p>Experimentally informed codon models (ExpCM) for a gene.</p>
<p>See <cite>__init__</cite> method for how to initialize an <cite>ExpCM</cite>.</p>
<p>Attributes should <strong>only</strong> be updated via the <cite>updateParams</cite>
method, do <strong>not</strong> set attributes directly.</p>
<dl class="simple">
<dt>Attributes (see also those inherited from <cite>Model</cite>):</dt><dd><dl class="simple">
<dt><cite>pi</cite> (<cite>numpy.ndarray</cite> of floats, shape <cite>(nsites, N_AA)</cite>)</dt><dd><p><cite>pi[r][a]</cite> is preference of site <cite>r</cite> for amino-acid <cite>a</cite>.</p>
</dd>
<dt><cite>kappa</cite> (float &gt; 0)</dt><dd><p>Transition-transversion ratio.</p>
</dd>
<dt><cite>omega</cite> (float &gt; 0)</dt><dd><p>Nonsynonymous to synonymous substitution ratio.</p>
</dd>
<dt><cite>beta</cite> (float &gt; 0)</dt><dd><p>Stringency parameter re-scaling amino-acid preferences.</p>
</dd>
<dt><cite>phi</cite> (<cite>numpy.ndarray</cite> of floats, length <cite>N_NT</cite>)</dt><dd><p>Nucleotide frequency params summing to one, is
computed from <cite>eta</cite>. We use <cite>eta</cite> rather than <cite>phi</cite>
as the free parameter during optimization.</p>
</dd>
<dt><cite>eta</cite> (<cite>numpy.ndarray</cite> of floats, length <cite>N_NT - 1</cite>)</dt><dd><p>Transformation of the nucleotide frequency params in <cite>phi</cite>;
all entries are &gt; 0 and &lt; 1. We use <cite>eta</cite> rather than <cite>phi</cite>
as the free parameter during optimization.</p>
</dd>
<dt><cite>Prxy</cite> (<cite>numpy.ndarray</cite> of floats, shape <cite>(nsites, N_CODON, N_CODON)</cite></dt><dd><p><cite>Prxy[r][x][y]</cite> is substitution rate from codon <cite>x</cite> to <cite>y</cite> at <cite>r</cite>.
Diagonal elements make rows sum to zero for each <cite>Prxy[r]</cite>.</p>
</dd>
<dt><cite>prx</cite> (<cite>numpy.ndarray</cite> of floats, shape <cite>(nsites, N_CODON)</cite></dt><dd><p><cite>prx[r][x]</cite> is stationary state of <cite>Prxy</cite> for codon <cite>x</cite> at <cite>r</cite>.
This attribute is equivalent to <cite>stationarystate</cite>.</p>
</dd>
<dt><cite>Qxy</cite> (<cite>numpy.ndarray</cite> of floats, shape <cite>(N_CODON, N_CODON)</cite></dt><dd><p><cite>Qxy[x][y]</cite> is mutation rate from <cite>x</cite> to <cite>y</cite>, diagonal undefined.</p>
</dd>
<dt><cite>Frxy</cite> (<cite>numpy.ndarray</cite> of floats, shape <cite>(nsites, N_CODON, N_CODON)</cite></dt><dd><p><cite>Frxy[r][x][y]</cite> fixation prob from <cite>x</cite> to <cite>y</cite>, diagonal
undefined for each <cite>Frxy[r]</cite>.</p>
</dd>
<dt><cite>Frxy_no_omega</cite></dt><dd><p>Like <cite>Frxy</cite> but <strong>not</strong> multiplied by <cite>omega</cite> for non-synonymous
mutations</p>
</dd>
<dt><cite>pi_codon</cite> (<cite>numpy.ndarray</cite> of floats, shape <cite>(nsites, N_CODON)</cite>)</dt><dd><p><cite>pi_codon[r][x]</cite> is preference of site <cite>r</cite> for amino acid
encoded by codon <cite>x</cite>.</p>
</dd>
<dt><cite>ln_pi_codon</cite> (<cite>numpy.ndarray</cite> floats, shape <cite>(nsites, N_CODON)</cite>)</dt><dd><p>Natural logarithm of <cite>pi_codon</cite>.</p>
</dd>
<dt><cite>piAx_piAy</cite> (<cite>numpy.ndarray</cite> floats, shape <cite>(nsites, N_CODON, N_CODON)</cite>)</dt><dd><p><cite>piAx_piAy[r][x][y]</cite> is ratio of preference for amino acid
encoded by codon <cite>x</cite> divided by pref for that encoded by <cite>y</cite>.</p>
</dd>
<dt><cite>piAx_piAy_beta</cite> (<cite>numpy.ndarray</cite> floats, shape <cite>(nsites, N_CODON, N_CODON)</cite>)</dt><dd><p>Equal to <cite>piAx_piAy</cite> raised to the power of <cite>beta</cite>.</p>
</dd>
<dt><cite>ln_piAx_piAy_beta (`numpy.ndarray</cite> of floats)</dt><dd><p>Natural logarithm of <cite>piAx_piAy_beta</cite>.</p>
</dd>
<dt><cite>dPrxy</cite> (dict)</dt><dd><p>Keyed by each string in <cite>freeparams</cite>, each value is <cite>numpy.ndarray</cite>
of floats giving derivative of <cite>Prxy</cite> with respect to that parameter.
The shape of each array <cite>(nsites, N_CODON, N_CODON)</cite> except for
<em>eta</em>, for which it is <cite>(N_NT - 1, nsites, N_CODON, N_CODON)</cite> with
the first index ranging over each element in <cite>eta</cite>.</p>
</dd>
<dt><cite>dprx</cite> (dict)</dt><dd><p>Keyed by strings in <cite>freeparams</cite>, each value is <cite>numpy.ndarray</cite>
of floats giving derivative of <cite>prx</cite> with respect to that param,
or 0 if if <cite>prx</cite> does not depend on parameter. The shape of each
array is <cite>(nsites, N_CODON)</cite> except for <em>eta</em>, for which it is
<cite>(N_NT - 1, nsites, N_CODON)</cite> with the first index ranging over
each element in <cite>eta</cite>. Equivalent to <cite>dstationarystate[param]</cite>.</p>
</dd>
<dt><cite>D</cite> (<cite>numpy.ndarray</cite> floats, shape <cite>(nsites, N_CODON)</cite>)</dt><dd><p>Element <cite>r</cite> is diagonal of <span class="math notranslate nohighlight">\(\mathbf{D_r}\)</span> diagonal matrix,
<span class="math notranslate nohighlight">\(\mathbf{P_r} = \mathbf{A_r}^{-1} \mathbf{D_r} \mathbf{A_r}\)</span></p>
</dd>
<dt><cite>A</cite> (<cite>numpy.ndarray</cite> floats, shape <cite>(nsites, N_CODON, N_CODON)</cite>)</dt><dd><p>Element r is <span class="math notranslate nohighlight">\(\mathbf{A_r}\)</span> matrix.</p>
</dd>
<dt><cite>Ainv</cite> (<cite>numpy.ndarray</cite> floats, shape <cite>(nsites, N_CODON, N_CODON)</cite>)</dt><dd><p>Element r is <span class="math notranslate nohighlight">\(\mathbf{A_r}^{-1}\)</span> matrix.</p>
</dd>
<dt><cite>B</cite> (dict)</dt><dd><p>Keyed by strings in <cite>freeparams</cite>, each value is <cite>numpy.ndarray</cite>
of floats giving <cite>B</cite> matrix for that parameter. The shape
is <cite>(nsites, N_CODON, N_CODON)</cite> except for <em>eta</em>, for which
it is <cite>(N_NT - 1, nsites, N_CODON, N_CODON)</cite> with the first
index ranging over each element in <cite>eta</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="attribute">
<dt id="phydmslib.models.ExpCM.ALLOWEDPARAMS">
<code class="sig-name descname">ALLOWEDPARAMS</code><em class="property"> = ['kappa', 'omega', 'beta', 'eta', 'mu']</em><a class="headerlink" href="#phydmslib.models.ExpCM.ALLOWEDPARAMS" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.M">
<code class="sig-name descname">M</code><span class="sig-paren">(</span><em class="sig-param">t</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM.M" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for method in <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.PARAMLIMITS">
<em class="property">property </em><code class="sig-name descname">PARAMLIMITS</code><a class="headerlink" href="#phydmslib.models.ExpCM.PARAMLIMITS" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="attribute">
<dt id="phydmslib.models.ExpCM.PARAMTYPES">
<code class="sig-name descname">PARAMTYPES</code><em class="property"> = {'beta': &lt;class 'float'&gt;, 'eta': (&lt;class 'numpy.ndarray'&gt;, (3,)), 'kappa': &lt;class 'float'&gt;, 'mu': &lt;class 'float'&gt;, 'omega': &lt;class 'float'&gt;, 'phi': (&lt;class 'numpy.ndarray'&gt;, (4,)), 'pi': &lt;class 'float'&gt;}</em><a class="headerlink" href="#phydmslib.models.ExpCM.PARAMTYPES" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">prefs, kappa=2.0, omega=0.5, beta=1.0, mu=1.0, phi=array([0.25, 0.25, 0.25, 0.25]), freeparams=['kappa', 'omega', 'beta', 'mu', 'eta']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize an <cite>ExpCM</cite> object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefs</cite> (list)</dt><dd><p>List of dicts giving amino-acid preferences for
each site. Each dict keyed by amino acid letter
codes, value is pref &gt; 0 and &lt; 1. Must sum to 1
at each site.</p>
</dd>
<dt><cite>kappa</cite>, <cite>omega</cite>, <cite>beta</cite>, <cite>mu</cite>, <cite>phi</cite></dt><dd><p>Model params described in main class doc string.</p>
</dd>
<dt><cite>freeparams</cite> (list of strings)</dt><dd><p>Specifies free parameters.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.branchScale">
<em class="property">property </em><code class="sig-name descname">branchScale</code><a class="headerlink" href="#phydmslib.models.ExpCM.branchScale" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.dM">
<code class="sig-name descname">dM</code><span class="sig-paren">(</span><em class="sig-param">t</em>, <em class="sig-param">param</em>, <em class="sig-param">Mt</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM.dM" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for method in <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.dlogprior">
<code class="sig-name descname">dlogprior</code><span class="sig-paren">(</span><em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM.dlogprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Zero for all <cite>param</cite>, as no prior defined over this model.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.dstationarystate">
<code class="sig-name descname">dstationarystate</code><span class="sig-paren">(</span><em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM.dstationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.freeparams">
<em class="property">property </em><code class="sig-name descname">freeparams</code><a class="headerlink" href="#phydmslib.models.ExpCM.freeparams" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.logprior">
<em class="property">property </em><code class="sig-name descname">logprior</code><a class="headerlink" href="#phydmslib.models.ExpCM.logprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Is zero, as no prior is defined over this model.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.mu">
<em class="property">property </em><code class="sig-name descname">mu</code><a class="headerlink" href="#phydmslib.models.ExpCM.mu" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.nsites">
<em class="property">property </em><code class="sig-name descname">nsites</code><a class="headerlink" href="#phydmslib.models.ExpCM.nsites" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.paramsReport">
<em class="property">property </em><code class="sig-name descname">paramsReport</code><a class="headerlink" href="#phydmslib.models.ExpCM.paramsReport" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.spielman_wr">
<code class="sig-name descname">spielman_wr</code><span class="sig-paren">(</span><em class="sig-param">norm=True</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM.spielman_wr" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of site-specific omega values calculated from the <cite>ExpCM</cite>.</p>
<blockquote>
<div><dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>norm</cite> (bool)</dt><dd><p>If <cite>True</cite>, normalize the <cite>omega_r</cite> values by the ExpCM
gene-wide <cite>omega</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>wr</cite> (list)</dt><dd><p>list of <cite>omega_r</cite> values of length <cite>nsites</cite></p>
</dd>
</dl>
</dd>
</dl>
</div></blockquote>
<p>Following
<a class="reference external" href="https://doi.org/10.1093/molbev/msv003">Spielman and Wilke, MBE, 32:1097-1108</a>,
we can predict the <cite>dN/dS</cite> value for each site <cite>r</cite>,
<span class="math notranslate nohighlight">\(\rm{spielman}\omega_r\)</span>, from the <cite>ExpCM</cite>.</p>
<p>When <cite>norm</cite> is <cite>False</cite>, the <cite>omega_r</cite> values are defined as
<span class="math notranslate nohighlight">\(\rm{spielman}\omega_r = \frac{\sum_x \sum_{y \in N_x}p_{r,x}        P_{r,xy}}{\sum_x \sum_{y \in Nx}p_{r,x}Q_{xy}}\)</span>,
where <cite>r,x,y</cite>, <span class="math notranslate nohighlight">\(p_{r,x}\)</span>, <span class="math notranslate nohighlight">\(P_{r,xy}\)</span>, and <span class="math notranslate nohighlight">\(Q_{x,y}\)</span>
have the same definitions as in the main <cite>ExpCM</cite> doc string and <span class="math notranslate nohighlight">\(N_{x}\)</span>
is the set of codons which are non-synonymous to codon <cite>x</cite> and differ from
<cite>x</cite> by one nucleotide.</p>
<p>When <cite>norm</cite> is <cite>True</cite>, the <cite>omega_r</cite> values above are divided by the
ExpCM <cite>omega</cite> value.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.stationarystate">
<em class="property">property </em><code class="sig-name descname">stationarystate</code><a class="headerlink" href="#phydmslib.models.ExpCM.stationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM.updateParams">
<code class="sig-name descname">updateParams</code><span class="sig-paren">(</span><em class="sig-param">newvalues</em>, <em class="sig-param">update_all=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM.updateParams" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.ExpCM_empirical_phi">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">ExpCM_empirical_phi</code><span class="sig-paren">(</span><em class="sig-param">prefs, g, kappa=2.0, omega=0.5, beta=1.0, mu=1.0, freeparams=['kappa', 'omega', 'beta', 'mu']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM_empirical_phi" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.ExpCM" title="phydmslib.models.ExpCM"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.ExpCM</span></code></a></p>
<p>An <cite>ExpCM</cite> with <cite>phi</cite> calculated empirically from nucleotide frequencies.</p>
<p>See <cite>__init__</cite> method for how to initialize an <cite>ExpCM_empirical_phi</cite>.</p>
<p>The difference between and <cite>ExpCM</cite> and an <cite>ExpCM_empirical_phi</cite> is that
the nucleotide frequency parameters <cite>phi</cite> are now calculated empirically
from their observed frequences (denoted <cite>g</cite>) in the codon alignment such
that the stationary state gives the empirically observed nucleotide
frequencies.</p>
<p>So compared to an <cite>ExpCM</cite>, <cite>eta</cite> (the transformation of <cite>phi</cite>) is no
longer a free parameter, but is computed as a function of other
model attributes.</p>
<dl class="simple">
<dt>Has all the attributes of an <cite>ExpCM</cite> plus the following:</dt><dd><dl class="simple">
<dt><cite>g</cite> (<cite>numpy.ndarray</cite> of float, length <cite>N_NT</cite>)</dt><dd><p>Empirical nucleotide frequencies in alignment, with
<cite>g[m]</cite> being frequency of nucleotide <cite>m</cite>. Must sum
to one.</p>
</dd>
<dt><cite>dphi_dbeta</cite> (<cite>numpy.ndarray</cite> of float, length <cite>N_NT</cite>)</dt><dd><p>Derivative of <cite>phi</cite> with respect to <cite>beta</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="attribute">
<dt id="phydmslib.models.ExpCM_empirical_phi.ALLOWEDPARAMS">
<code class="sig-name descname">ALLOWEDPARAMS</code><em class="property"> = ['kappa', 'omega', 'beta', 'mu']</em><a class="headerlink" href="#phydmslib.models.ExpCM_empirical_phi.ALLOWEDPARAMS" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="phydmslib.models.ExpCM_empirical_phi.PARAMTYPES">
<code class="sig-name descname">PARAMTYPES</code><em class="property"> = {'beta': &lt;class 'float'&gt;, 'eta': (&lt;class 'numpy.ndarray'&gt;, (3,)), 'g': (&lt;class 'numpy.ndarray'&gt;, (4,)), 'kappa': &lt;class 'float'&gt;, 'mu': &lt;class 'float'&gt;, 'omega': &lt;class 'float'&gt;, 'phi': (&lt;class 'numpy.ndarray'&gt;, (4,)), 'pi': &lt;class 'float'&gt;}</em><a class="headerlink" href="#phydmslib.models.ExpCM_empirical_phi.PARAMTYPES" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM_empirical_phi.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">prefs, g, kappa=2.0, omega=0.5, beta=1.0, mu=1.0, freeparams=['kappa', 'omega', 'beta', 'mu']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM_empirical_phi.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize an <cite>ExpCM_empirical_phi</cite> object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefs</cite>, <cite>kappa</cite>, <cite>omega</cite>, <cite>beta</cite>, <cite>mu</cite>, <cite>freeparams</cite></dt><dd><p>Same meaning as for an <cite>ExpCM</cite></p>
</dd>
<dt><cite>g</cite></dt><dd><p>Has the meaning described in the main class doc string.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.ExpCM_empirical_phi_divpressure">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">ExpCM_empirical_phi_divpressure</code><span class="sig-paren">(</span><em class="sig-param">prefs, g, divPressureValues, kappa=2.0, omega=0.5, beta=1.0, mu=1.0, omega2=0.0, freeparams=['kappa', 'omega', 'beta', 'mu', 'omega2']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM_empirical_phi_divpressure" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.ExpCM_empirical_phi" title="phydmslib.models.ExpCM_empirical_phi"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.ExpCM_empirical_phi</span></code></a></p>
<p><cite>ExpCM</cite> with empirical <cite>phi</cite> and <em>a priori</em> diversifying pressure at sites.</p>
<p>See <cite>__init__</cite> method for how to initialize.</p>
<p>Difference between <cite>ExpCM_empirical_phi</cite> and <cite>ExpCM_empirical_phi_divpressure</cite>
is that <cite>omega</cite> is replaced by <cite>omega * (1 + omega2 * deltar)</cite> where
<cite>deltar</cite> is the expectation of diversifying pressure at site <cite>r</cite>.</p>
<p>The rate of non-synonymous to synonymous change is proportional to the
expectation of diversifying pressure.</p>
<dl class="simple">
<dt>Has all the attributes of an <cite>ExpCM_empirical_phi</cite> plus the following:</dt><dd><dl class="simple">
<dt><cite>omega2</cite> (float)</dt><dd><p>Part of the expression which replaces <cite>omega</cite>.</p>
</dd>
<dt><cite>divPressureValues</cite> (<cite>numpy.array</cite> of length <cite>nsites</cite>)</dt><dd><p>Array of <cite>deltar</cite> values in site order.
The maximum absolute value of the array should be between 0 and 1.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="attribute">
<dt id="phydmslib.models.ExpCM_empirical_phi_divpressure.ALLOWEDPARAMS">
<code class="sig-name descname">ALLOWEDPARAMS</code><em class="property"> = ['kappa', 'omega', 'beta', 'mu', 'omega2']</em><a class="headerlink" href="#phydmslib.models.ExpCM_empirical_phi_divpressure.ALLOWEDPARAMS" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="phydmslib.models.ExpCM_empirical_phi_divpressure.PARAMTYPES">
<code class="sig-name descname">PARAMTYPES</code><em class="property"> = {'beta': &lt;class 'float'&gt;, 'eta': (&lt;class 'numpy.ndarray'&gt;, (3,)), 'g': (&lt;class 'numpy.ndarray'&gt;, (4,)), 'kappa': &lt;class 'float'&gt;, 'mu': &lt;class 'float'&gt;, 'omega': &lt;class 'float'&gt;, 'omega2': &lt;class 'float'&gt;, 'phi': (&lt;class 'numpy.ndarray'&gt;, (4,)), 'pi': &lt;class 'float'&gt;}</em><a class="headerlink" href="#phydmslib.models.ExpCM_empirical_phi_divpressure.PARAMTYPES" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM_empirical_phi_divpressure.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">prefs, g, divPressureValues, kappa=2.0, omega=0.5, beta=1.0, mu=1.0, omega2=0.0, freeparams=['kappa', 'omega', 'beta', 'mu', 'omega2']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM_empirical_phi_divpressure.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize an <cite>ExpCM_empirical_phi_divpressure</cite> object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>prefs</cite>, <cite>kappa</cite>, <cite>omega</cite>, <cite>beta</cite>, <cite>mu</cite>, <cite>g</cite>, <cite>freeparams</cite></dt><dd><p>Same meaning as for an <cite>ExpCM_empirical_phi</cite></p>
</dd>
<dt><cite>divPressureValues</cite>, <cite>omega2</cite></dt><dd><p>Meaning described in the main class doc string.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.ExpCM_fitprefs">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">ExpCM_fitprefs</code><span class="sig-paren">(</span><em class="sig-param">prefs, prior, kappa, omega, phi, mu=1.0, origbeta=1.0, freeparams=['zeta']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM_fitprefs" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.ExpCM" title="phydmslib.models.ExpCM"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.ExpCM</span></code></a></p>
<p>An <cite>ExpCM</cite> with the preferences <cite>pi</cite> as free parameters.</p>
<p>The difference between <cite>ExpCM</cite> and <cite>ExpCM_fitprefs</cite> is that the
amino-acid preferences <cite>pi</cite> are optimized for the latter. There
can also be a regularizing prior on these preferences.</p>
<p>The way this class is currently implemented, it will be inefficient
if there is more than one site. The recommended usage is to optimize
across-site parameters with fixed preferences, then optimize preferences
for each site individually with this class after fixing across-site
parameters. You currently can <em>not</em> initialize with more than one
site.</p>
<p>See <cite>__init__</cite> method for how to initialize an <cite>ExpCM_fitprefs</cite>.</p>
<dl class="simple">
<dt>Has all the attributes of an <cite>ExpCM</cite>, plus:</dt><dd><dl class="simple">
<dt><cite>zeta</cite> (<cite>numpy.ndarray</cite> of float, shape <cite>(nsites * (N_AA - 1))</cite>)</dt><dd><p>Transformation of the preferences <cite>pi</cite>; all entries are
&gt; 0 and &lt; 1. We use <cite>zeta</cite> rather than <cite>pi</cite> as the free
parameter during optimization. The entry
<cite>zeta.reshape(nsites, N_AA - 1)[r][i]</cite> is value <cite>i</cite>
for site <cite>r</cite> (<cite>0 &lt;= r &lt; nsites</cite>) where <cite>0 &lt;= i &lt; N_AA - 1</cite>.</p>
</dd>
<dt><cite>dPrxy</cite> (dict)</dt><dd><p>Like for <cite>ExpCM</cite>, but also contains entry keyed by ‘zeta’ of
of shape <cite>(nsites * (N_AA - 1), nsites, N_CODON, N_CODON)</cite>
where the first index ranges over the elements in <cite>zeta</cite>.</p>
</dd>
<dt><cite>tildeFrxy</cite> (<cite>numpy.ndarray</cite> of <cite>float</cite>, shape <cite>(nsites, N_CODON, N_CODON)</cite>)</dt><dd><p>Contains quantities used in calculating derivative of
<cite>dPrxy</cite> with respect to <cite>zeta</cite>.</p>
</dd>
<dt><cite>origpi</cite> (<cite>numpy.ndarray</cite> of float, shape <cite>(nsites, N_AA)</cite>)</dt><dd><p><cite>origpi[r][a]</cite> is the original preference for amino-acid
<cite>a</cite> at site <cite>r</cite> as given by the calling parameter <cite>prefs</cite>
and re-scaled by <cite>origbeta</cite>. This value is <strong>not</strong> updated
as <cite>zeta</cite> and <cite>pi</cite> are optimized, and is rather used to
keep track of where the optimization starts. Set to one
if you have not already optimized a stringency parameter
prior to initializing this class.</p>
</dd>
<dt><cite>prior</cite> (<cite>None</cite> or a tuple)</dt><dd><p>Specifies the regularizing prior over the preferences.
This prior is centered on the preferences in <cite>origpi</cite>.
A value of <cite>None</cite> indicates no regularizing prior.
A tuple of <cite>(‘invquadratic’, c1, c2)</cite> is the inverse
quadratic prior described in Bloom, <em>Biology Direct</em>, 12:1.</p>
</dd>
</dl>
</dd>
</dl>
<p><cite>beta</cite> (the stringency parameter) is <strong>not</strong> a free parameter,
and is instead fixed to 1. This is because it does not make sense
to optimize a stringency parameter if you are also optimizing the
preferences as they are confounded.</p>
<dl class="attribute">
<dt id="phydmslib.models.ExpCM_fitprefs.ALLOWEDPARAMS">
<code class="sig-name descname">ALLOWEDPARAMS</code><em class="property"> = ['kappa', 'omega', 'eta', 'mu', 'zeta']</em><a class="headerlink" href="#phydmslib.models.ExpCM_fitprefs.ALLOWEDPARAMS" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM_fitprefs.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">prefs, prior, kappa, omega, phi, mu=1.0, origbeta=1.0, freeparams=['zeta']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM_fitprefs.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize an <cite>ExpCM_fitprefs</cite> object.</p>
<p>The calling parameters have the meaning described in the main class doc
string. The default values have changed consistent with recommended
usage of this class: that you have already estimated
the across-site parameters (<cite>kappa</cite>, <cite>omega</cite>, <cite>mu</cite>, <cite>phi</cite>) and so
are specifying fixed values for those. You are then just optimizing the
preferences in their variable-transformed form <cite>zeta</cite> from the initial
values in <cite>prefs</cite> scaled by <cite>origbeta</cite>. The meaning of <cite>origbeta</cite>
is described in the main class doc string.</p>
<dl class="simple">
<dt>Additional arguments:</dt><dd><dl class="simple">
<dt><cite>origbeta</cite> (<cite>float</cite>)</dt><dd><p>The preferences in <cite>prefs</cite> are re-scaled by <cite>origbeta</cite> and
used as the initial value and stored in <cite>origpi</cite>. <cite>origbeta</cite>
might differ from 1 if you already optimized <cite>prefs</cite> using
a fixed-preference <cite>ExpCM</cite> prior to initializing this model.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM_fitprefs.dlogprior">
<code class="sig-name descname">dlogprior</code><span class="sig-paren">(</span><em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM_fitprefs.dlogprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Value of derivative of prior depends on value of <cite>prior</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.ExpCM_fitprefs.logprior">
<em class="property">property </em><code class="sig-name descname">logprior</code><a class="headerlink" href="#phydmslib.models.ExpCM_fitprefs.logprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Value of log prior depends on value of <cite>prior</cite>.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.ExpCM_fitprefs2">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">ExpCM_fitprefs2</code><span class="sig-paren">(</span><em class="sig-param">prefs, prior, kappa, omega, phi, mu=1.0, origbeta=1.0, freeparams=['zeta']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.ExpCM_fitprefs2" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.ExpCM_fitprefs" title="phydmslib.models.ExpCM_fitprefs"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.ExpCM_fitprefs</span></code></a></p>
<p>An <cite>ExpCM</cite> with the preferences <cite>pi</cite> as free parameters.</p>
<p>This class differs from <cite>ExpCM_fitprefs</cite> in the way that
it internally handles the transformation from the preferences
<cite>pi</cite> to the optimization variables <cite>zeta</cite>. All other attributes
are the same as for an <cite>ExpCM_fitprefs</cite>.</p>
</dd></dl>

<dl class="class">
<dt id="phydmslib.models.GammaDistributedBetaModel">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">GammaDistributedBetaModel</code><span class="sig-paren">(</span><em class="sig-param">model, ncats, alpha_lambda=1.0, beta_lambda=2.0, freeparams=['alpha_lambda', 'beta_lambda']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedBetaModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.GammaDistributedModel" title="phydmslib.models.GammaDistributedModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.GammaDistributedModel</span></code></a></p>
<p>Implements gamma distribution over <cite>beta</cite> for a model.</p>
<p>See <cite>__init__</cite> method for how to initialize a
<cite>GammaDistributedBetaModel</cite>.</p>
<p>Attributes should <strong>only</strong> be updated via the <cite>updateParams</cite>
method; do <strong>not</strong> set attributes directly.</p>
<p>Attributes are inherited from <cite>GammaDistributedModel</cite>.</p>
<dl class="method">
<dt id="phydmslib.models.GammaDistributedBetaModel.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">model, ncats, alpha_lambda=1.0, beta_lambda=2.0, freeparams=['alpha_lambda', 'beta_lambda']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedBetaModel.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize an <cite>GammaDistributedModel</cite> object.</p>
<p>The <cite>lambda_param</cite> is set to “beta”.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>model</cite> <cite>ncats</cite>,`alpha_lambda`, <cite>beta_lambda</cite>, <cite>freeparams</cite></dt><dd><p>Meaning described in main class doc string for
<cite>GammaDistributedModel</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.GammaDistributedModel">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">GammaDistributedModel</code><span class="sig-paren">(</span><em class="sig-param">model, lambda_param, ncats, alpha_lambda=1.0, beta_lambda=2.0, freeparams=['alpha_lambda', 'beta_lambda']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.DistributionModel" title="phydmslib.models.DistributionModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.DistributionModel</span></code></a></p>
<p>Implements gamma distribution over some parameter <cite>lambda</cite> for a model.</p>
<p>This model can be used to take a simple substitution model that
directly subclasses <cite>Model</cite> and implement a gamma distribution
over <em>one</em> parameter <cite>lambda</cite>. For instance, if this is done for
the <cite>YNGKP_M0</cite> model parameter <cite>omega</cite>, it yields a M5 variant of
the YNGKP model.</p>
<p>The <cite>lambda</cite> parameter is drawn from <cite>ncats</cite> categories, with
the values being at the mean of each of these categories and
equal weight assigned to each category.</p>
<p>This means that rather than optimizing <cite>lambda</cite> directly,
we optimize the shape and inverse-scale parameters
of its gamma distribution.</p>
<p>See <cite>__init__</cite> method for how to initialize a
<cite>GammaDistributedModel</cite>.</p>
<p>Attributes should <strong>only</strong> be updated via the <cite>updateParams</cite>
method; do <strong>not</strong> set attributes directly.</p>
<dl class="simple">
<dt>Attributes (see also those inherited from <cite>DistributionModel</cite>):</dt><dd><dl class="simple">
<dt><cite>alpha_omega</cite> (<cite>float</cite> &gt; 0)</dt><dd><p>Gamma distribution shape parameter.</p>
</dd>
<dt><cite>beta_omega</cite> (<cite>float</cite> &gt; 0)</dt><dd><p>Gamma distribution inverse-scale parameter</p>
</dd>
<dt><cite>lambda_param</cite> (str)</dt><dd><p>Parameter which the gamma distirubtion is implemented over.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.M">
<code class="sig-name descname">M</code><span class="sig-paren">(</span><em class="sig-param">k</em>, <em class="sig-param">t</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.M" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>DistributionModel</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.PARAMLIMITS">
<em class="property">property </em><code class="sig-name descname">PARAMLIMITS</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.PARAMLIMITS" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="attribute">
<dt id="phydmslib.models.GammaDistributedModel.PARAMTYPES">
<code class="sig-name descname">PARAMTYPES</code><em class="property"> = {'alpha_lambda': &lt;class 'float'&gt;, 'beta_lambda': &lt;class 'float'&gt;}</em><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.PARAMTYPES" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">model, lambda_param, ncats, alpha_lambda=1.0, beta_lambda=2.0, freeparams=['alpha_lambda', 'beta_lambda']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a <cite>GammaDistributedModel</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>model</cite> (<cite>Model</cite>)</dt><dd><p>A substitution model with <cite>distributedparam</cite> as a free parameter
that does <strong>not</strong> have any other property drawn from
a distribution. Any other parameter that is in <cite>freeparams</cite>
of this model will also be an optimized parameter.</p>
</dd>
<dt><cite>ncats</cite>, <cite>lambda_param</cite>, <cite>alpha_omega</cite>, <cite>beta_omega</cite></dt><dd><p>Meaning described in main class doc string.</p>
</dd>
<dt><cite>freeparams</cite></dt><dd><p>The free parameters will be these <strong>plus</strong> anything
in <cite>model.freeparams</cite> other than <cite>distributedparam</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.basemodel">
<em class="property">property </em><code class="sig-name descname">basemodel</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.basemodel" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>DistributionModel</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.branchScale">
<em class="property">property </em><code class="sig-name descname">branchScale</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.branchScale" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.catweights">
<em class="property">property </em><code class="sig-name descname">catweights</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.catweights" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>DistributionModel</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.dM">
<code class="sig-name descname">dM</code><span class="sig-paren">(</span><em class="sig-param">k</em>, <em class="sig-param">t</em>, <em class="sig-param">param</em>, <em class="sig-param">Mkt</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.dM" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>DistributionModel</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.d_distributionparams">
<em class="property">property </em><code class="sig-name descname">d_distributionparams</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.d_distributionparams" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>DistributionModel</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.distributedparam">
<em class="property">property </em><code class="sig-name descname">distributedparam</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.distributedparam" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns name of the distributed parameter.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.distributionparams">
<em class="property">property </em><code class="sig-name descname">distributionparams</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.distributionparams" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns list of params defining distribution of <cite>distributedparam</cite>.</p>
<p>This list is <cite>[‘alpha_lambda’, ‘beta_lambda’]</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.dlogprior">
<code class="sig-name descname">dlogprior</code><span class="sig-paren">(</span><em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.dlogprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Equal to value of <cite>basemodel.dlogprior</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.dstationarystate">
<code class="sig-name descname">dstationarystate</code><span class="sig-paren">(</span><em class="sig-param">k</em>, <em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.dstationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.freeparams">
<em class="property">property </em><code class="sig-name descname">freeparams</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.freeparams" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.logprior">
<em class="property">property </em><code class="sig-name descname">logprior</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.logprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Equal to value of <cite>basemodel.logprior</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.mu">
<em class="property">property </em><code class="sig-name descname">mu</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.mu" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.ncats">
<em class="property">property </em><code class="sig-name descname">ncats</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.ncats" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>DistributionModel</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.nsites">
<em class="property">property </em><code class="sig-name descname">nsites</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.nsites" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.paramsReport">
<em class="property">property </em><code class="sig-name descname">paramsReport</code><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.paramsReport" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.stationarystate">
<code class="sig-name descname">stationarystate</code><span class="sig-paren">(</span><em class="sig-param">k</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.stationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.GammaDistributedModel.updateParams">
<code class="sig-name descname">updateParams</code><span class="sig-paren">(</span><em class="sig-param">newvalues</em>, <em class="sig-param">update_all=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedModel.updateParams" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.GammaDistributedOmegaModel">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">GammaDistributedOmegaModel</code><span class="sig-paren">(</span><em class="sig-param">model, ncats, alpha_lambda=1.0, beta_lambda=2.0, freeparams=['alpha_lambda', 'beta_lambda']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedOmegaModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.GammaDistributedModel" title="phydmslib.models.GammaDistributedModel"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.GammaDistributedModel</span></code></a></p>
<p>Implements gamma distribution over <cite>omega</cite> for a model.</p>
<p>This model can be used to implement a gamma distribution over a
<cite>Models</cite> <cite>omega</cite> parameter. For instance, if this is done for the
<cite>YNGKP_M0</cite> model, it yields a M5 variant of the YNGKP model.</p>
<p>See <cite>__init__</cite> method for how to initialize a
<cite>GammaDistributedOmegaModel</cite>.</p>
<p>Attributes should <strong>only</strong> be updated via the <cite>updateParams</cite>
method; do <strong>not</strong> set attributes directly.</p>
<p>Attributes are inherited from <cite>GammaDistributedModel</cite>.</p>
<dl class="method">
<dt id="phydmslib.models.GammaDistributedOmegaModel.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">model, ncats, alpha_lambda=1.0, beta_lambda=2.0, freeparams=['alpha_lambda', 'beta_lambda']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.GammaDistributedOmegaModel.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize an <cite>GammaDistributedModel</cite> object.</p>
<p>The <cite>lambda_param</cite> is set to “omega”.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>model</cite> <cite>ncats</cite>,`alpha_lambda`, <cite>beta_lambda</cite>, <cite>freeparams</cite></dt><dd><p>Meaning described in main class doc string for
<cite>GammaDistributedModel</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.Model">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">Model</code><a class="headerlink" href="#phydmslib.models.Model" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Substitution model abstract base class.</p>
<p>Specifies required methods / attributes of substitution models.</p>
<dl class="method">
<dt id="phydmslib.models.Model.M">
<em class="property">abstract </em><code class="sig-name descname">M</code><span class="sig-paren">(</span><em class="sig-param">t</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.Model.M" title="Permalink to this definition">¶</a></dt>
<dd><p>Matrix exponential <cite>M(mu * t) = exp(mu * t * P)</cite>.</p>
<p>If we are considering branch to tip node, you can use
<cite>tips</cite> and <cite>gaps</cite> to get only the column needed for
likelihood calculation. This is computationally cheaper.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>t</cite> (float &gt; 0)</dt><dd><p>The branch length.</p>
</dd>
<dt><cite>tips</cite> (<cite>None</cite> or <cite>numpy.ndarray</cite> of <cite>int</cite>, length <cite>nsites</cite>)</dt><dd><p>If <cite>None</cite>, return the full matrix exponential.
Otherwise, should be an array with element <cite>r</cite>
giving the codon at tip node <cite>r</cite> (just put 0
if codon is a gap and see <cite>gaps</cite> below).</p>
</dd>
<dt><cite>gaps</cite> (<cite>None</cite> or <cite>numpy.ndarray</cite> of <cite>int</cite>)</dt><dd><p>Only meaningful if using <cite>tips</cite>. In this case,
array should list <cite>r</cite> for all sites where tip
node codon is a gap. <cite>None</cite> is equivalent to
no gaps (also can be empty array).</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>M</cite> (<cite>numpy.ndarray</cite> of floats)</dt><dd><dl class="simple">
<dt>If not using <cite>tips</cite>, shape is <cite>(nsites, N_CODON, N_CODON)</cite></dt><dd><p>with <cite>M(t)[r][x][y]</cite> probability <cite>r</cite> changes from <cite>x</cite>
to <cite>y</cite> in time <cite>t</cite>.</p>
</dd>
<dt>If using <cite>tips</cite>, shape is <cite>(nsites, N_CODON)</cite> with</dt><dd><p><cite>M(t)[r][x] probability r changes from `x</cite> to <cite>tips[r]</cite>
in time <cite>t</cite>. If <cite>r</cite> is in <cite>gap</cite>, then <cite>M(t)[r][x]</cite> is
one.</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.PARAMLIMITS">
<em class="property">abstract property </em><code class="sig-name descname">PARAMLIMITS</code><a class="headerlink" href="#phydmslib.models.Model.PARAMLIMITS" title="Permalink to this definition">¶</a></dt>
<dd><p>Dict of tuples giving lower and upper allowed param values.</p>
<p>For each <cite>param</cite> in <cite>freeparams</cite>, <cite>PARAMLIMITS[param]</cite>
is a 2-tuple of floats giving upper and lower allowed
values of <cite>param</cite> (or of each element in <cite>param</cite> if
<cite>param</cite> is an array).</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.branchScale">
<em class="property">abstract property </em><code class="sig-name descname">branchScale</code><a class="headerlink" href="#phydmslib.models.Model.branchScale" title="Permalink to this definition">¶</a></dt>
<dd><p>Factor to scale branch lengths to substitutions per site.</p>
<p>When we use a branch length of <cite>t = 1</cite>, how many substitutions
per site do we expect averaged over all sites? This method
returns that number. So to scale branch lengths to units of
substitution per site, multiply by this number.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.dM">
<em class="property">abstract </em><code class="sig-name descname">dM</code><span class="sig-paren">(</span><em class="sig-param">t</em>, <em class="sig-param">param</em>, <em class="sig-param">Mt</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.Model.dM" title="Permalink to this definition">¶</a></dt>
<dd><p>Derivative of <cite>M(t)</cite> with respect to <cite>param</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>t</cite> (float &gt; 0)</dt><dd><p>The branch length.</p>
</dd>
<dt><cite>param</cite> (string in <cite>freeparams</cite> or the string <cite>‘t’</cite>)</dt><dd><p>Differentiate with respect to this model parameter,
or with respect to branch length if <cite>‘t’</cite>.</p>
</dd>
<dt><cite>Mt</cite> (<cite>numpy.ndarray</cite>.)</dt><dd><p>The current value of <cite>M(t, tips, gaps)</cite>. Typically this
has already been pre-computed which is why it is passed
here to avoid computing again. Otherwise pass
<cite>None</cite> and it will be re-computed.</p>
</dd>
<dt><cite>tips</cite> and <cite>gaps</cite></dt><dd><p>Same meaning as for <cite>M</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>dM_param</cite> (<cite>numpy.ndarray</cite> floats)</dt><dd><p>If <cite>param</cite> is a float, then <cite>dM_param[r][x][y]</cite>
is derivative of <cite>M(t)[r][x][y]</cite> with respect to
<cite>param</cite>. If <cite>param</cite> is an array, then
<cite>dM_param[i][r][x][y]</cite> is derivative of <cite>M(t)[r][x][y]</cite>
with respect to <cite>param[i]</cite>. If using <cite>tips</cite>, then
<cite>dM_param[r]</cite> or <cite>dM_params[i][r]</cite> is only a
single array providing the derivative of
<cite>M(t, tips, gaps)[r]</cite> with respect to <cite>param</cite> or
<cite>param[i]</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.dlogprior">
<em class="property">abstract </em><code class="sig-name descname">dlogprior</code><span class="sig-paren">(</span><em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.Model.dlogprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Derivative of <cite>logprior</cite> with respect to <cite>param</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><p><cite>param</cite> (string in <cite>freeparams</cite>)</p>
</dd>
<dt>Returns:</dt><dd><p>Derivative of <cite>logprior</cite> with respect to <cite>param</cite>,
or 0 if there is no prior defined over this model.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.dstationarystate">
<em class="property">abstract </em><code class="sig-name descname">dstationarystate</code><span class="sig-paren">(</span><em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.Model.dstationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>Derivative of <cite>stationarystate</cite> with respect to <cite>param</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><p><cite>param</cite> (string in <cite>freeparams</cite>)</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>dstationarystate</cite> (<cite>numpy.ndarray</cite> of floats or zero)</dt><dd><p>If <cite>param</cite> is a float, then <cite>dstationarystate[r][x]</cite>
is derivative of <cite>stationarystate[r][x]</cite> with respect
to <cite>param</cite>. If <cite>param</cite> is an array, then
<cite>dstationarystate[i][r][x]</cite> is derivative of
<cite>stationarystate[r][x]</cite> with respect to <cite>param[i]</cite>.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.freeparams">
<em class="property">abstract property </em><code class="sig-name descname">freeparams</code><a class="headerlink" href="#phydmslib.models.Model.freeparams" title="Permalink to this definition">¶</a></dt>
<dd><p>Model parameters to be optimized (a list of strings).</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.logprior">
<em class="property">abstract property </em><code class="sig-name descname">logprior</code><a class="headerlink" href="#phydmslib.models.Model.logprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Log prior over current model.</p>
<p>Is a float giving log prior over current model, or
0 if there is no prior defined over this model.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.mu">
<em class="property">abstract property </em><code class="sig-name descname">mu</code><a class="headerlink" href="#phydmslib.models.Model.mu" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutation rate, scales branch lengths in <cite>M</cite> and <cite>dM</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.nsites">
<em class="property">abstract property </em><code class="sig-name descname">nsites</code><a class="headerlink" href="#phydmslib.models.Model.nsites" title="Permalink to this definition">¶</a></dt>
<dd><p>An <cite>int</cite> giving the number of codon sites.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.paramsReport">
<em class="property">abstract property </em><code class="sig-name descname">paramsReport</code><a class="headerlink" href="#phydmslib.models.Model.paramsReport" title="Permalink to this definition">¶</a></dt>
<dd><p>Reports current values of independent model parameters.</p>
<p>Returns a dictionary keyed by parameter name with value being the
parameter value. Does <strong>not</strong> include <cite>mu</cite> as this is confounded
with branch lengths.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.stationarystate">
<em class="property">abstract property </em><code class="sig-name descname">stationarystate</code><a class="headerlink" href="#phydmslib.models.Model.stationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>Stationary state of substitution model.</p>
<p>A <cite>numpy.ndarray</cite> of floats, shape <cite>(nsites, N_CODON)</cite>.
Element <cite>stationarystate[r][x]</cite> is the stationary
state probability of codon <cite>x</cite> at site <cite>r</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.Model.updateParams">
<em class="property">abstract </em><code class="sig-name descname">updateParams</code><span class="sig-paren">(</span><em class="sig-param">newvalues</em>, <em class="sig-param">update_all=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.Model.updateParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Update model params.</p>
<p>This method is the <strong>only</strong> acceptable way to update
model parameters; other dependent model parameters are
automatically updated as needed if you use this method.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>newvalues</cite> (dict)</dt><dd><p>Can be keyed by any parameter in <cite>freeparams</cite>.</p>
</dd>
<dt><cite>update_all</cite> (bool)</dt><dd><p>If <cite>True</cite>, update all dependent attributes using
current values of model parameters.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="phydmslib.models.YNGKP_M0">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.models.</code><code class="sig-name descname">YNGKP_M0</code><span class="sig-paren">(</span><em class="sig-param">e_pw, nsites, kappa=2.0, omega=0.5, mu=1.0, freeparams=['kappa', 'omega', 'mu']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.YNGKP_M0" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#phydmslib.models.Model" title="phydmslib.models.Model"><code class="xref py py-class docutils literal notranslate"><span class="pre">phydmslib.models.Model</span></code></a></p>
<p>YNGKP_M0 model from Yang et al, 2000.</p>
<p>See <cite>__init__</cite> method for how to initialize an <cite>YNGKP_M0</cite>.</p>
<p>Attributes should <strong>only</strong> be updated via the <cite>updateParams</cite>
method, do <strong>not</strong> set attributes directly.</p>
<dl class="simple">
<dt>Attributes (see also those inherited from <cite>Model</cite>):</dt><dd><dl class="simple">
<dt><cite>kappa</cite> (float &gt; 0)</dt><dd><p>Transition-transversion ratio.</p>
</dd>
<dt><cite>omega</cite> (float &gt; 0)</dt><dd><p>Nonsynonymous to synonymous substitution ratio.</p>
</dd>
<dt><cite>Pxy</cite> (<cite>numpy.ndarray</cite> of floats, shape <cite>(1, CODON, N_CODON)</cite></dt><dd><p><cite>Pxy[0][x][y]</cite> is substitution rate from codon <cite>x</cite> to <cite>y</cite>.
Diagonal elements make rows sum to zero.</p>
</dd>
<dt><cite>Pxy_no_omega</cite></dt><dd><p>Like <cite>Pxy</cite> but <strong>not</strong> multiplied by <cite>omega</cite> for nonsynonymous
mutations.</p>
</dd>
<dt><cite>dPxy</cite> (dict)</dt><dd><p>Keyed by each string in <cite>freeparams</cite>, each value is <cite>numpy.ndarray</cite>
of floats giving derivative of <cite>Pxy</cite> with respect to that parameter.
The shape of each array <cite>(1, N_CODON, N_CODON)</cite>.</p>
</dd>
<dt><cite>e_pw</cite> (scipy.ndarray, shape <cite>(3, N_NT)</cite>)</dt><dd><p>The empirical nucleotide frequencies for each position in a codon
measured from the alignment. <cite>e_pw[p][w]</cite> give the frequency of
nucleotide <cite>w</cite> at codon position <cite>p</cite>.</p>
</dd>
<dt><cite>Phi_x</cite> (scipy.ndarray, shape <cite>(N_CODON,)</cite>)</dt><dd><p>The codon frequencies calculated from the <cite>phi</cite> frequencies.
<cite>Phi_x[x]</cite> = phi[0][x0] * phi[1][x1] * phi[2][x2]</p>
</dd>
<dt><cite>phi</cite> (scipy.ndarray, shape <cite>(3, N_NT)</cite>)</dt><dd><p>The model nucleotide frequencies for each position in a codon
computed using the <cite>CF3X4</cite> estimator. <cite>phi[p][w]</cite> is the model
frequency of nucleotide <cite>w</cite> at codon position <cite>p</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<p>Unlike the <cite>ExpCM</cite>, <cite>YNGKP_M0</cite> does not have site-specific calculations.
In order to maintain consistency, most attributes, such as <cite>Pxy</cite> and <cite>dPxy</cite>
do have a single “site” dimension which is carried through the calculations.
When <cite>M</cite> and <cite>dM</cite> are returned, the single site matrix is repeated so the
final dimensions of <cite>M</cite> and <cite>dM</cite> are match those in the docs for the <cite>Models</cite>
abstract base class.</p>
<p>The model nucleotide frequences, <cite>phi</cite> are computed using the
empirical nucleotide frequencies <cite>e_pw</cite> and the <cite>Corrected F3X4</cite>
estimator (Pond et al, 2010). Unlike the traditional <cite>MG3X4</cite> or <cite>GY3X4</cite>,
this estimator takes into account stop codon nucleotide frequencies.
The codon frequencies, <cite>Phi_x</cite> are computed from the <cite>phi</cite> values.</p>
<dl class="attribute">
<dt id="phydmslib.models.YNGKP_M0.ALLOWEDPARAMS">
<code class="sig-name descname">ALLOWEDPARAMS</code><em class="property"> = ['kappa', 'omega', 'mu']</em><a class="headerlink" href="#phydmslib.models.YNGKP_M0.ALLOWEDPARAMS" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.M">
<code class="sig-name descname">M</code><span class="sig-paren">(</span><em class="sig-param">t</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.YNGKP_M0.M" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for method in <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.PARAMLIMITS">
<em class="property">property </em><code class="sig-name descname">PARAMLIMITS</code><a class="headerlink" href="#phydmslib.models.YNGKP_M0.PARAMLIMITS" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="attribute">
<dt id="phydmslib.models.YNGKP_M0.PARAMTYPES">
<code class="sig-name descname">PARAMTYPES</code><em class="property"> = {'e_pw': (&lt;class 'numpy.ndarray'&gt;, (3, 4)), 'kappa': &lt;class 'float'&gt;, 'mu': &lt;class 'float'&gt;, 'omega': &lt;class 'float'&gt;}</em><a class="headerlink" href="#phydmslib.models.YNGKP_M0.PARAMTYPES" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">e_pw, nsites, kappa=2.0, omega=0.5, mu=1.0, freeparams=['kappa', 'omega', 'mu']</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.YNGKP_M0.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize an <cite>YNGKP_M0</cite> object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>kappa</cite>, <cite>omega</cite>, <cite>mu</cite>,</dt><dd><p>Model params described in main class doc string.</p>
</dd>
<dt><cite>freeparams</cite> (list of strings)</dt><dd><p>Specifies free parameters.</p>
</dd>
<dt><cite>e_pw</cite>, <cite>nsites</cite></dt><dd><p>Meaning described in the main class doc string.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.branchScale">
<em class="property">property </em><code class="sig-name descname">branchScale</code><a class="headerlink" href="#phydmslib.models.YNGKP_M0.branchScale" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.dM">
<code class="sig-name descname">dM</code><span class="sig-paren">(</span><em class="sig-param">t</em>, <em class="sig-param">param</em>, <em class="sig-param">Mt</em>, <em class="sig-param">tips=None</em>, <em class="sig-param">gaps=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.YNGKP_M0.dM" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for method in <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.dlogprior">
<code class="sig-name descname">dlogprior</code><span class="sig-paren">(</span><em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.YNGKP_M0.dlogprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Zero for all <cite>param</cite>, as no prior defined over this model.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.dstationarystate">
<code class="sig-name descname">dstationarystate</code><span class="sig-paren">(</span><em class="sig-param">param</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.YNGKP_M0.dstationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.freeparams">
<em class="property">property </em><code class="sig-name descname">freeparams</code><a class="headerlink" href="#phydmslib.models.YNGKP_M0.freeparams" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.logprior">
<em class="property">property </em><code class="sig-name descname">logprior</code><a class="headerlink" href="#phydmslib.models.YNGKP_M0.logprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Is zero, as no prior is defined over this model.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.mu">
<em class="property">property </em><code class="sig-name descname">mu</code><a class="headerlink" href="#phydmslib.models.YNGKP_M0.mu" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.nsites">
<em class="property">property </em><code class="sig-name descname">nsites</code><a class="headerlink" href="#phydmslib.models.YNGKP_M0.nsites" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.paramsReport">
<em class="property">property </em><code class="sig-name descname">paramsReport</code><a class="headerlink" href="#phydmslib.models.YNGKP_M0.paramsReport" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.stationarystate">
<em class="property">property </em><code class="sig-name descname">stationarystate</code><a class="headerlink" href="#phydmslib.models.YNGKP_M0.stationarystate" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.models.YNGKP_M0.updateParams">
<code class="sig-name descname">updateParams</code><span class="sig-paren">(</span><em class="sig-param">newvalues</em>, <em class="sig-param">update_all=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.models.YNGKP_M0.updateParams" title="Permalink to this definition">¶</a></dt>
<dd><p>See docs for <cite>Model</cite> abstract base class.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-phydmslib.numutils">
<span id="phydmslib-numutils-module"></span><h2>phydmslib.numutils module<a class="headerlink" href="#module-phydmslib.numutils" title="Permalink to this headline">¶</a></h2>
<p>Numerical routines for <code class="docutils literal notranslate"><span class="pre">phydmslib</span></code>.</p>
<dl class="function">
<dt id="phydmslib.numutils.broadcastGetCols">
<code class="sig-prename descclassname">phydmslib.numutils.</code><code class="sig-name descname">broadcastGetCols</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.numutils.broadcastGetCols" title="Permalink to this definition">¶</a></dt>
<dd><p>Get specified columns from <cite>ndarray</cite> of square <cite>ndarrays</cite>.</p>
<p>This functions uses fast <cite>numpy</cite> broadcasting to get the
columns from a long array of arrays.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>m</cite> (<cite>numpy.ndarray</cite>, shape <cite>(r, n, n)</cite></dt><dd><p>Array of matrices from which we get the columns</p>
</dd>
<dt><cite>cols</cite> (<cite>numpy.ndarray</cite>, type <cite>int</cite>, length <cite>r</cite>)</dt><dd><p>All entries should be &gt;= 0 and &lt; <cite>n</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>mcols</cite> (<cite>numpy.ndarray</cite>, shape <cite>(r, n)</cite>)</dt><dd><p><cite>mcols[r]</cite> is equal to <cite>mcols[r][cols[r]]</cite></p>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="mi">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">r</span> <span class="o">*</span> <span class="n">n</span> <span class="o">*</span> <span class="n">n</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cols</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random_integers</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">n</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">expected</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">m</span><span class="p">[</span><span class="n">i</span><span class="p">][:,</span> <span class="n">cols</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">r</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scipy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">expected</span><span class="p">,</span> <span class="n">broadcastGetCols</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">cols</span><span class="p">))</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="phydmslib.numutils.broadcastMatrixMultiply">
<code class="sig-prename descclassname">phydmslib.numutils.</code><code class="sig-name descname">broadcastMatrixMultiply</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.numutils.broadcastMatrixMultiply" title="Permalink to this definition">¶</a></dt>
<dd><p>Broadcast square matrix multiplication using <cite>blas</cite> for speed.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>a</cite> (<cite>numpy.ndarray</cite>, shape <cite>(r, n, n)</cite>)</dt><dd><p>First matrix.</p>
</dd>
<dt><cite>b</cite> (<cite>numpy.ndarray</cite>, shape <cite>(r, n, n)</cite>)</dt><dd><p>Second matrix.</p>
</dd>
<dt><cite>alpha</cite> (<cite>float</cite>)</dt><dd><p>Multiply each entry in product by this (same meaning
as for BLAS <cite>dgemm</cite>).</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>ab</cite> (<cite>numpy.ndarray</cite>, shape <cite>(r, n, n)</cite>)</dt><dd><p><cite>ab[r]</cite> is the matrix product of <cite>a[r]</cite>
with <cite>b[r]</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="mi">3</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">r</span> <span class="o">*</span> <span class="n">n</span> <span class="o">*</span> <span class="n">n</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;float&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">b</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">r</span> <span class="o">*</span> <span class="n">n</span> <span class="o">*</span> <span class="n">n</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;float&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">n</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">r</span><span class="p">):</span>
<span class="gp">... </span>  <span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ab</span> <span class="o">=</span> <span class="n">broadcastMatrixMultiply</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">expected</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">matmul</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scipy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">expected</span><span class="p">,</span> <span class="n">ab</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="phydmslib.numutils.broadcastMatrixVectorMultiply">
<code class="sig-prename descclassname">phydmslib.numutils.</code><code class="sig-name descname">broadcastMatrixVectorMultiply</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.numutils.broadcastMatrixVectorMultiply" title="Permalink to this definition">¶</a></dt>
<dd><p>Broadcast matrix vector multiplication using <cite>blas</cite> for speed.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>m</cite> (<cite>numpy.ndarray</cite>, shape <cite>(d1, d2, d2)</cite>)</dt><dd><p>Array of square matrices to multiply.</p>
</dd>
<dt><cite>v</cite> (<cite>numpy.ndarray</cite>, shape <cite>(d1, d2)</cite>)</dt><dd><p>Array of vectors to multiply.</p>
</dd>
<dt><cite>alpha</cite> (<cite>float</cite>)</dt><dd><p>Multiply each entry in product by this (same meaning
as for BLAS <cite>dgemv</cite>).</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>mv</cite> (<cite>numpy.ndarray</cite>, shape <cite>(d1, d2)</cite>)</dt><dd><p><cite>mv[r]</cite> is the matrix-vector product of <cite>m[r]</cite> with
<cite>v[r]</cite> for 0 &lt;= <cite>r</cite> &lt;= <cite>d1</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([[[</span><span class="mf">1.</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]],</span> <span class="p">[[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span> <span class="p">[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">]],</span> <span class="p">[[</span><span class="mi">9</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="p">[</span><span class="mi">7</span><span class="p">,</span> <span class="mi">6</span><span class="p">]]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">v</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">1.</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mv</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">ndarray</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;int&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
<span class="gp">... </span>  <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
<span class="gp">... </span>     <span class="n">mvrx</span> <span class="o">=</span> <span class="mi">0</span>
<span class="gp">... </span>     <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
<span class="gp">... </span>       <span class="n">mvrx</span> <span class="o">+=</span> <span class="n">m</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="n">x</span><span class="p">][</span><span class="n">y</span><span class="p">]</span> <span class="o">*</span> <span class="n">v</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="n">y</span><span class="p">]</span>
<span class="gp">... </span>     <span class="n">mv</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="n">x</span><span class="p">]</span> <span class="o">=</span> <span class="n">mvrx</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mv2</span> <span class="o">=</span> <span class="n">broadcastMatrixVectorMultiply</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mv</span><span class="o">.</span><span class="n">shape</span> <span class="o">==</span> <span class="n">mv2</span><span class="o">.</span><span class="n">shape</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scipy</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">mv</span><span class="p">,</span> <span class="n">mv2</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-phydmslib.parsearguments">
<span id="phydmslib-parsearguments-module"></span><h2>phydmslib.parsearguments module<a class="headerlink" href="#module-phydmslib.parsearguments" title="Permalink to this headline">¶</a></h2>
<p>Module for parsing arguments.</p>
<dl class="class">
<dt id="phydmslib.parsearguments.ArgumentDefaultsRawDescriptionFormatter">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">ArgumentDefaultsRawDescriptionFormatter</code><span class="sig-paren">(</span><em class="sig-param">prog</em>, <em class="sig-param">indent_increment=2</em>, <em class="sig-param">max_help_position=24</em>, <em class="sig-param">width=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.ArgumentDefaultsRawDescriptionFormatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.ArgumentDefaultsHelpFormatter</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.RawDescriptionHelpFormatter</span></code></p>
<p>Print default arguments and raw text description formatter.</p>
<p>Based on this:
<a class="reference external" href="http://stackoverflow.com/questions/18462610/argumentparser-epilog-and-description-formatting-in-conjunction-with-argumentdef">http://stackoverflow.com/questions/18462610/argumentparser-epilog-and-description-formatting-in-conjunction-with-argumentdef</a></p>
</dd></dl>

<dl class="class">
<dt id="phydmslib.parsearguments.ArgumentParserNoArgHelp">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">ArgumentParserNoArgHelp</code><span class="sig-paren">(</span><em class="sig-param">prog=None</em>, <em class="sig-param">usage=None</em>, <em class="sig-param">description=None</em>, <em class="sig-param">epilog=None</em>, <em class="sig-param">parents=[]</em>, <em class="sig-param">formatter_class=&lt;class 'argparse.HelpFormatter'&gt;</em>, <em class="sig-param">prefix_chars='-'</em>, <em class="sig-param">fromfile_prefix_chars=None</em>, <em class="sig-param">argument_default=None</em>, <em class="sig-param">conflict_handler='error'</em>, <em class="sig-param">add_help=True</em>, <em class="sig-param">allow_abbrev=True</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.ArgumentParserNoArgHelp" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">argparse.ArgumentParser</span></code></p>
<p>Like <em>argparse.ArgumentParser</em>, but prints help when no arguments.</p>
<dl class="method">
<dt id="phydmslib.parsearguments.ArgumentParserNoArgHelp.error">
<code class="sig-name descname">error</code><span class="sig-paren">(</span><em class="sig-param">message</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.ArgumentParserNoArgHelp.error" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints error message, then help.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.ExistingFile">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">ExistingFile</code><span class="sig-paren">(</span><em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.ExistingFile" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>fname</em> is name of an existing file return it, otherwise an error.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.ExistingFileOrNone">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">ExistingFileOrNone</code><span class="sig-paren">(</span><em class="sig-param">fname</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.ExistingFileOrNone" title="Permalink to this definition">¶</a></dt>
<dd><p>Like <cite>Existingfile</cite>, but if <cite>fname</cite> is string “None” then return <cite>None</cite>.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.FloatBetweenZeroAndOne">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">FloatBetweenZeroAndOne</code><span class="sig-paren">(</span><em class="sig-param">x</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.FloatBetweenZeroAndOne" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>x</em> only if <em>0 &lt;= x &lt;= 1</em>, otherwise raises error.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.FloatGreaterThanEqualToZero">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">FloatGreaterThanEqualToZero</code><span class="sig-paren">(</span><em class="sig-param">x</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.FloatGreaterThanEqualToZero" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>x</em> is a float &gt;= 0, returns it, otherwise raises and error.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%.1f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">FloatGreaterThanEqualToZero</span><span class="p">(</span><span class="s1">&#39;1.5&#39;</span><span class="p">))</span>
<span class="go">1.5</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%.1f</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">FloatGreaterThanEqualToZero</span><span class="p">(</span><span class="s1">&#39;-1.1&#39;</span><span class="p">))</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">-1.1 not float greater than or equal to zero</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.FloatGreaterThanOne">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">FloatGreaterThanOne</code><span class="sig-paren">(</span><em class="sig-param">x</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.FloatGreaterThanOne" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>x</em> is a string for a float &gt; 1, returns it, otherwise an error.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.FloatGreaterThanZero">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">FloatGreaterThanZero</code><span class="sig-paren">(</span><em class="sig-param">x</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.FloatGreaterThanZero" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>x</em> is string for float &gt; 0, returns it, otherwise an error.</p>
<p>Designed based on this: <a class="reference external" href="http://stackoverflow.com/questions/12116685/how-can-i-require-my-python-scripts-argument-to-be-a-float-between-0-0-1-0-usin">http://stackoverflow.com/questions/12116685/how-can-i-require-my-python-scripts-argument-to-be-a-float-between-0-0-1-0-usin</a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%.3f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">FloatGreaterThanZero</span><span class="p">(</span><span class="s1">&#39;0.1&#39;</span><span class="p">))</span>
<span class="go">0.100</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">FloatGreaterThanZero</span><span class="p">(</span><span class="s1">&#39;0.0&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">0.0 not a float greater than zero</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.IntGreaterThanOne">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">IntGreaterThanOne</code><span class="sig-paren">(</span><em class="sig-param">n</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.IntGreaterThanOne" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>n</em> is an integer &gt; 1, returns it, otherwise an error.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.IntGreaterThanZero">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">IntGreaterThanZero</code><span class="sig-paren">(</span><em class="sig-param">n</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.IntGreaterThanZero" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>n</em> is an integer &gt; 0, returns it, otherwise an error.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.ModelOption">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">ModelOption</code><span class="sig-paren">(</span><em class="sig-param">model</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.ModelOption" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>model</em> if a valid choice.</p>
<p>Returns the string if it specifies a <code class="docutils literal notranslate"><span class="pre">YNGKP_</span></code> model variant.</p>
<p>Returns <em>(‘ExpCM’, prefsfile)</em> if it specifies an <code class="docutils literal notranslate"><span class="pre">ExpCM_</span></code> model.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.NonNegativeInt">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">NonNegativeInt</code><span class="sig-paren">(</span><em class="sig-param">n</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.NonNegativeInt" title="Permalink to this definition">¶</a></dt>
<dd><p>If <em>n</em> is non-negative integer returns it, otherwise an error.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">NonNegativeInt</span><span class="p">(</span><span class="s1">&#39;8&#39;</span><span class="p">))</span>
<span class="go">8</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">NonNegativeInt</span><span class="p">(</span><span class="s1">&#39;8.1&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">8.1 is not an integer</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">NonNegativeInt</span><span class="p">(</span><span class="s1">&#39;0&#39;</span><span class="p">))</span>
<span class="go">0</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">NonNegativeInt</span><span class="p">(</span><span class="s1">&#39;-1&#39;</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
   <span class="c">...</span>
<span class="gr">ValueError</span>: <span class="n">-1 is not non-negative</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.PhyDMSComprehensiveParser">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">PhyDMSComprehensiveParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.PhyDMSComprehensiveParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <code class="docutils literal notranslate"><span class="pre">phdyms_comprehensive</span></code> script.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.PhyDMSLogoPlotParser">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">PhyDMSLogoPlotParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.PhyDMSLogoPlotParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <cite>argparse.ArgumentParser</cite> for <code class="docutils literal notranslate"><span class="pre">phydms_logoplot</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.PhyDMSParser">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">PhyDMSParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.PhyDMSParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <code class="docutils literal notranslate"><span class="pre">phydms</span></code> script.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.PhyDMSPrepAlignmentParser">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">PhyDMSPrepAlignmentParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.PhyDMSPrepAlignmentParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <code class="docutils literal notranslate"><span class="pre">phydms_prepalignment</span></code>.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.PhyDMSTestdivpressureParser">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">PhyDMSTestdivpressureParser</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.PhyDMSTestdivpressureParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns <em>argparse.ArgumentParser</em> for <code class="docutils literal notranslate"><span class="pre">phdyms_testdivpressure</span></code>
script.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.parsearguments.diffPrefsPrior">
<code class="sig-prename descclassname">phydmslib.parsearguments.</code><code class="sig-name descname">diffPrefsPrior</code><span class="sig-paren">(</span><em class="sig-param">priorstring</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.parsearguments.diffPrefsPrior" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses <cite>priorstring</cite> and returns <cite>prior</cite> tuple.</p>
</dd></dl>

</div>
<div class="section" id="module-phydmslib.simulate">
<span id="phydmslib-simulate-module"></span><h2>phydmslib.simulate module<a class="headerlink" href="#module-phydmslib.simulate" title="Permalink to this headline">¶</a></h2>
<p>Functions for performing simulations, mostly using <code class="docutils literal notranslate"><span class="pre">pyvolve</span></code>.</p>
<p>Written by Jesse Bloom and Sarah Hilton.</p>
<dl class="function">
<dt id="phydmslib.simulate.pyvolvePartitions">
<code class="sig-prename descclassname">phydmslib.simulate.</code><code class="sig-name descname">pyvolvePartitions</code><span class="sig-paren">(</span><em class="sig-param">model</em>, <em class="sig-param">divselection=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.simulate.pyvolvePartitions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get list of <cite>pyvolve</cite> partitions for <cite>model</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>model</cite> (<cite>phydmslib.models.Models</cite> object)</dt><dd><p>The model used for the simulations. Currently only
certain <cite>Models</cite> are supported (e.g., <cite>YNGKP</cite>,
<cite>ExpCM</cite>)</p>
</dd>
<dt><cite>divselection</cite> (<cite>None</cite> or 2-tuple <cite>(divomega, divsites)</cite>)</dt><dd><p>Set this option if you want to simulate a subset of sites
as under diversifying selection (e.g., an <cite>omega</cite> different
than that used by <cite>model</cite>. In this case, <cite>divomega</cite> is
the omega for this subset of sites, and <cite>divsites</cite> is a list
of the sites in 1, 2, … numbering.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt><cite>partitions</cite> (<cite>list</cite> of <cite>pyvolve.Partition</cite> objects)</dt><dd><p>Can be fed into <cite>pyvolve.Evolver</cite> to simulate evolution.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="phydmslib.simulate.simulateAlignment">
<code class="sig-prename descclassname">phydmslib.simulate.</code><code class="sig-name descname">simulateAlignment</code><span class="sig-paren">(</span><em class="sig-param">model</em>, <em class="sig-param">treeFile</em>, <em class="sig-param">alignmentPrefix</em>, <em class="sig-param">randomSeed=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.simulate.simulateAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate an alignment given a model and tree (units = subs/site).</p>
<p>Simulations done using <cite>pyvolve</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>model</cite> (<cite>phydmslib.models.Models</cite> object)</dt><dd><p>The model used for the simulations. Only
models that can be passed to <cite>pyvolve.Partitions</cite>
are supported.</p>
</dd>
<dt><cite>treeFile</cite> (str)</dt><dd><p>Name of newick file used to simulate the sequences.
The branch lengths should be in substitutions per site,
which is the default units for all <cite>phydms</cite> outputs.</p>
</dd>
<dt><cite>alignmentPrefix</cite></dt><dd><p>Prefix for the files created by <cite>pyvolve</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<p>The result of this function is a simulated FASTA alignment
file with the name having the prefix giving by <cite>alignmentPrefix</cite>
and the suffix <cite>‘_simulatedalignment.fasta’</cite>.</p>
</dd></dl>

</div>
<div class="section" id="module-phydmslib.treelikelihood">
<span id="phydmslib-treelikelihood-module"></span><h2>phydmslib.treelikelihood module<a class="headerlink" href="#module-phydmslib.treelikelihood" title="Permalink to this headline">¶</a></h2>
<p>Tree likelihoods.</p>
<p>Nucleotides, amino acids, and codons are indexed by integers 0, 1, …
using the indexing schemes defined in <cite>phydmslib.constants</cite>.</p>
<dl class="class">
<dt id="phydmslib.treelikelihood.TreeLikelihood">
<em class="property">class </em><code class="sig-prename descclassname">phydmslib.treelikelihood.</code><code class="sig-name descname">TreeLikelihood</code><span class="sig-paren">(</span><em class="sig-param">tree</em>, <em class="sig-param">alignment</em>, <em class="sig-param">model</em>, <em class="sig-param">underflowfreq=3</em>, <em class="sig-param">dparamscurrent=True</em>, <em class="sig-param">dtcurrent=False</em>, <em class="sig-param">branchScale=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Uses alignment, model, and tree to calculate likelihoods.</p>
<p>See <cite>__init__</cite> for how to initialize a <cite>TreeLikelihood</cite>.</p>
<p>Most commonly, you will initiate a tree likelihood object, and
then maximize it using the <cite>maximizeLikelihood</cite> method.</p>
<p>A <cite>TreeLikelihood</cite> object is designed to only work for a fixed
tree topology. In the internal workings, this fixed topology is
transformed into the node indexing scheme defined by <cite>name_to_nodeindex</cite>.</p>
<p>After initialization, most attributes should <strong>not</strong> be changed directly.
They should only be altered via the <cite>updateParams</cite> method or
by altering the <cite>paramsarray</cite> property. Both of these methods
of updating attributes correctly update all of the dependent
properties; setting other attributes directly does not.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><dl class="simple">
<dt><cite>tree</cite> (instance of <cite>Bio.Phylo.BaseTree.Tree</cite> derived class)</dt><dd><p>Phylogenetic tree. Branch lengths are in units of codon
substitutions per site for the current <cite>model</cite>.</p>
</dd>
<dt><cite>model</cite> (instance of <cite>phydmslib.models.Model</cite> derived class)</dt><dd><p>Specifies the substitution model for <cite>nsites</cite> codon sites.
This can either be a simple <cite>Model</cite> or a <cite>DistributionModel</cite>.
Note that if a non-zero prior is defined by <cite>model</cite>, this
is included in <cite>loglik</cite> and <cite>dloglik</cite>.</p>
</dd>
<dt><cite>alignment</cite> (list of 2-tuples of strings, <cite>(head, seq)</cite>)</dt><dd><p>Aligned protein-coding codon sequences. Headers match
tip names in <cite>tree</cite>; sequences contain <cite>nsites</cite> codons.</p>
</dd>
<dt><cite>paramsarray</cite> (<cite>numpy.ndarray</cite> of floats, 1-dimensional)</dt><dd><p>An array of the model free parameters. You can directly
assign to <cite>paramsarray</cite>, and the correct parameters will
be internally updated vi <cite>updateParams</cite>.
<cite>paramsarray</cite> is actually a property rather than an attribute.</p>
</dd>
<dt><cite>paramsarraybounds</cite> (<cite>tuple</cite> of 2-tuples)</dt><dd><p><cite>paramsarraybounds[i]</cite> is the 2-tuple <cite>(minbound, maxbound)</cite>
for parameter <cite>paramsarray[i]</cite>. Set <cite>minbound</cite> or <cite>maxbound</cite>
are <cite>None</cite> if there is no lower or upper bound.</p>
</dd>
<dt><cite>loglik</cite> (<cite>float</cite>)</dt><dd><p>Current log likelihood. Also includes <cite>model.logprior</cite>
if this is non-zero.</p>
</dd>
<dt><cite>siteloglik</cite> (<cite>numpy.ndarray</cite> of floats, length <cite>nsites</cite>)</dt><dd><p><cite>siteloglik[r]</cite> is current log likelihood at site <cite>r</cite>.</p>
</dd>
<dt><cite>dparamscurrent</cite> (<cite>bool</cite>)</dt><dd><p>Do we keep the derivatives of the likelihood with
respect to model parameters current? Doing so
involves computational costs, so only set to <cite>True</cite>
if using these derivatives. Currently, only one of
<cite>dparamscurrent</cite> and <cite>dtcurrent</cite> can be <cite>True</cite> simultaneously,
as we optimize parameters and branch lengths separately.</p>
</dd>
<dt><cite>dtcurrent</cite> (<cite>bool</cite>)</dt><dd><p>Do we keep the derivatives of the likelihood with
respect to branch lengths current? Doing so
involves computational costs, so only set to <cite>True</cite>
if using these derivatives.</p>
</dd>
<dt><cite>dloglik</cite> (<cite>dict</cite>)</dt><dd><p>For each <cite>param</cite> in <cite>model.freeparams</cite>, <cite>dloglik[param]</cite>
is the derivative of <cite>loglik</cite> with respect to <cite>param</cite>.
Only accessible if <cite>dparamscurrent</cite> is <cite>True</cite>. Also
includes <cite>model.dlogprior</cite> if this is non-zero.</p>
</dd>
<dt><cite>dloglikarray</cite> (<cite>numpy.ndarray</cite> of floats, 1-dimensional)</dt><dd><p><cite>dloglikarray[i]</cite> is the derivative of <cite>loglik</cite> with respect
to the parameter represented in <cite>paramsarray[i]</cite>. This is
the same information as in <cite>dloglik</cite>, but in a different
representation. Only accessible if <cite>dparamscurrent</cite> is <cite>True</cite>.</p>
</dd>
<dt><cite>dloglik_dt</cite> (<cite>numpy.ndarray</cite> of floats, shape <cite>(nnodes - 1,)</cite>.</dt><dd><p><cite>dloglik_dt</cite> is the derivative of <cite>loglik</cite> with respect
to the branch lengths <cite>t</cite>.</p>
</dd>
<dt><cite>nsites</cite> (int)</dt><dd><p>Number of codon sites.</p>
</dd>
<dt><cite>nseqs</cite> (int)</dt><dd><p>Number of sequences in <cite>alignment</cite>, and tip nodes in <cite>tree</cite>.</p>
</dd>
<dt><cite>nnodes</cite> (int)</dt><dd><p>Total number of nodes in <cite>tree</cite>, counting tip and internal.
Node are indexed 0, 1, …, <cite>nnodes - 1</cite>. This indexing is done
such that the descendants of node <cite>n</cite> always have indices &lt; <cite>n</cite>,
and so that the tips are the first <cite>ntips</cite> indices and the
internals are the last <cite>ninternal</cite> indices.</p>
</dd>
<dt><cite>ninternal</cite> (int)</dt><dd><p>Total number of internal nodes in <cite>tree</cite>.</p>
</dd>
<dt><cite>ntips</cite> (int)</dt><dd><p>Total number of tip nodes in <cite>tree</cite>.</p>
</dd>
<dt><cite>tips</cite> (<cite>numpy.ndarray</cite> of <cite>int</cite>, shape <cite>(ntips, nsites)</cite>)</dt><dd><p><cite>tips[n][r]</cite> is the index of the codon at site <cite>r</cite>
for tip node <cite>n</cite> (0 &lt;= <cite>n</cite> &lt; <cite>ntips</cite>). If <cite>tips[n][r]</cite> is
a gap, then <cite>tips[n][r]</cite> is 0 and <cite>r</cite> is in <cite>gaps[n]</cite>.</p>
</dd>
<dt><cite>gaps</cite> (<cite>numpy.array</cite>, length <cite>ntips</cite>, entries <cite>numpy.array</cite> of <cite>int</cite>)</dt><dd><p><cite>gaps[n]</cite> is an array containing all sites where the codon is
a gap for tip node <cite>n</cite> (0 &lt;= <cite>n</cite> &lt; <cite>ntips</cite>).</p>
</dd>
<dt><cite>name_to_nodeindex</cite> (<cite>dict</cite>)</dt><dd><p>For each sequence name (these are headers in <cite>alignment</cite>,
tip names in <cite>tree</cite>), <cite>name_to_nodeindex[name]</cite> is the <cite>int</cite>
index of that <cite>node</cite> in the internal indexing scheme.</p>
</dd>
<dt><cite>rdescend</cite> and <cite>ldescend</cite> (lists of <cite>int</cite> of length <cite>ninternal</cite>)</dt><dd><p>For each internal node <cite>n</cite>, <cite>rdescend[n - ntips]</cite>
is its right descendant and <cite>ldescend[n - ntips]</cite> is its left
descendant.</p>
</dd>
<dt><cite>descendants</cite> (<cite>list</cite> of <cite>list</cite>)</dt><dd><p>For each node <cite>n</cite>, <cite>descendants[n]</cite> is a list of all nodes
that are direct or indirect descendants of <cite>n</cite>.</p>
</dd>
<dt><cite>t</cite> (<cite>numpy.ndarray</cite> of <cite>float</cite>, shape <cite>(nnodes - 1,)</cite>)</dt><dd><p><cite>t[n]</cite> is the branch length leading node <cite>n</cite>. The branch leading
to the root node (which has index <cite>nnodes - 1</cite>) is undefined
and not used, which is why <cite>t</cite> is of length <cite>nnodes - 1</cite> rather
than <cite>nnodes</cite>.</p>
</dd>
<dt><cite>L</cite> (<cite>dict</cite>)</dt><dd><p><cite>L[n]</cite> is a <cite>numpy.ndarray</cite> of <cite>float</cite> of shape <cite>(nsites, N_CODON)</cite>.
for each node <cite>n</cite>. <cite>L[n][r][x] is the partial condition likelihood
of codon `x</cite> at site <cite>r</cite> at node <cite>n</cite>. Note that these
must be corrected by adding <cite>underflowlogscale</cite>.
If <cite>model</cite> is a <cite>DistributionModel</cite>, then <cite>L[n]</cite> is instead
of shape <cite>(model.ncats, nsites, N_CODON)</cite> with the
first index ranging over the model categories.
Entries in <cite>L</cite> are only defined and saved for nodes as needed
in order to save memory.</p>
</dd>
<dt><cite>dL</cite> (<cite>dict</cite> keyed by strings)</dt><dd><p>For each free model parameter <cite>param</cite> and node <cite>n</cite>,
<cite>dL[param][n]</cite> is derivative of <cite>L[n]</cite> with respect to <cite>param</cite>.
Only guaranteed to be valid if <cite>dparamscurrent</cite> is <cite>True</cite>,
and entries are only defined and saved for nodes as needed
in order to save memory.</p>
</dd>
<dt><cite>dL_dt</cite> (<cite>dict</cite> of <cite>dict</cite>)</dt><dd><p><cite>dL_dt[n][n2]</cite> is the derivative of <cite>L[n2]</cite> with respect to
branch length <cite>t[n]</cite> where <cite>0 &lt;= n &lt; nnodes - 1</cite>.</p>
</dd>
<dt><cite>underflowfreq</cite> (<cite>int</cite> &gt;= 1)</dt><dd><p>The frequency with which we rescale likelihoods to avoid
numerical underflow</p>
</dd>
<dt><cite>underflowlogscale</cite> (<cite>numpy.ndarray</cite> of <cite>float</cite>, shape <cite>(nsites,)</cite>)</dt><dd><p>Corrections that must be added to <cite>L</cite> at the root node to
get actual likelihoods when underflow correction is being
performed.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">tree</em>, <em class="sig-param">alignment</em>, <em class="sig-param">model</em>, <em class="sig-param">underflowfreq=3</em>, <em class="sig-param">dparamscurrent=True</em>, <em class="sig-param">dtcurrent=False</em>, <em class="sig-param">branchScale=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a <cite>TreeLikelihood</cite> object.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>tree</cite>, <cite>alignment</cite>, <cite>model</cite>, <cite>underflowfreq</cite></dt><dd><p>Attributes of same name described in class doc string.
Note that we make copies of <cite>tree</cite>, <cite>model</cite>, and
<cite>alignment</cite>’ so the calling objects are not modified
during optimization.</p>
</dd>
<dt><cite>dparamscurrent</cite>, <cite>dtcurrent</cite></dt><dd><p>Current values of these flags indicating which
derivatives to compute.</p>
</dd>
<dt><cite>branchScale</cite> (<cite>None</cite> or float &gt; 0)</dt><dd><p>The branch lengths in the input tree are assumed
to be in units of substitutions per site with
the scaling defined by <cite>model.branchScale</cite>. If
the scaling should instead be defined by some
other value of <cite>branchScale</cite>, indicate by
setting to that value rather than <cite>None</cite>. This
is useful if tree was inferred on models across
many sites but you are now just analyzing an
individual one. Note that this option applies
only to the input tree, not the output one.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.dloglik">
<em class="property">property </em><code class="sig-name descname">dloglik</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.dloglik" title="Permalink to this definition">¶</a></dt>
<dd><p><cite>dloglik[param]</cite> is derivative of <cite>loglik</cite> by <cite>param</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.dloglik_dt">
<em class="property">property </em><code class="sig-name descname">dloglik_dt</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.dloglik_dt" title="Permalink to this definition">¶</a></dt>
<dd><p><cite>dloglik_dt</cite> is derivative of <cite>loglik</cite> with respect to <cite>t</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.dloglikarray">
<em class="property">property </em><code class="sig-name descname">dloglikarray</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.dloglikarray" title="Permalink to this definition">¶</a></dt>
<dd><p>Derivative of <cite>loglik</cite> with respect to <cite>paramsarray</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.dparamscurrent">
<em class="property">property </em><code class="sig-name descname">dparamscurrent</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.dparamscurrent" title="Permalink to this definition">¶</a></dt>
<dd><p>Are derivatives with respect to model parameters current?</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.dtcurrent">
<em class="property">property </em><code class="sig-name descname">dtcurrent</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.dtcurrent" title="Permalink to this definition">¶</a></dt>
<dd><p>Are derivatives with respect to branch lengths current?</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.maximizeLikelihood">
<code class="sig-name descname">maximizeLikelihood</code><span class="sig-paren">(</span><em class="sig-param">optimize_brlen=False</em>, <em class="sig-param">approx_grad=False</em>, <em class="sig-param">logliktol=0.01</em>, <em class="sig-param">nparamsretry=1</em>, <em class="sig-param">printfunc=None</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.maximizeLikelihood" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximize the log likelihood.</p>
<p>Maximizes log likelihood with respect to model parameters
and potentially branch lengths depending on <cite>optimize_brlen</cite>.
If optimizing the branch lengths, iterates between optimizing
the model parameters and branch lengths.</p>
<p>Uses the L-BFGS-B method implemented in <cite>scipy.optimize</cite>.</p>
<p>There is no return variable, but after call object attributes
will correspond to maximimum likelihood values.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>optimize_brlen</cite> (bool)</dt><dd><p>Do we optimize branch lengths?</p>
</dd>
<dt><cite>approx_grad</cite> (bool)</dt><dd><p>If <cite>True</cite>, then we numerically approximate the gradient
rather than using the analytical values.</p>
</dd>
<dt><cite>logliktol</cite> (float)</dt><dd><p>When using <cite>optimize_brlen</cite>, keep iterating between
optimization of parameters and branch lengths until
change in log likelihood is less than <cite>logliktol</cite>.</p>
</dd>
<dt><cite>nparamsretry</cite> (int &gt;= 0)</dt><dd><p>Number of times to retry parameter optimization from
different initial values if it fails the first time.</p>
</dd>
<dt><cite>printfunc</cite> (<cite>None</cite> or a function)</dt><dd><p>If not <cite>None</cite>, then we print using <cite>printfunc</cite> the
detailed results of the optimization at each step.
For instance, <cite>printfunc</cite> might be <cite>sys.stderr.write</cite>
or <cite>logger.info</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A string giving a summary of the maximization.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.paramsarray">
<em class="property">property </em><code class="sig-name descname">paramsarray</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.paramsarray" title="Permalink to this definition">¶</a></dt>
<dd><p>All free model parameters as 1-dimensional <cite>numpy.ndarray</cite>.</p>
<p>You are allowed to update model parameters by direct
assignment of this property.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.paramsarraybounds">
<em class="property">property </em><code class="sig-name descname">paramsarraybounds</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.paramsarraybounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Bounds for parameters in <cite>paramsarray</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.t">
<em class="property">property </em><code class="sig-name descname">t</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.t" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets array of branch lengths.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.tree">
<em class="property">property </em><code class="sig-name descname">tree</code><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Tree with branch lengths in codon substitutions per site.</p>
<p>The tree is a <cite>Bio.Phylo.BaseTree.Tree</cite> object.</p>
<p>This is the current tree after whatever optimizations have
been performed so far.</p>
</dd></dl>

<dl class="method">
<dt id="phydmslib.treelikelihood.TreeLikelihood.updateParams">
<code class="sig-name descname">updateParams</code><span class="sig-paren">(</span><em class="sig-param">newvalues</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.treelikelihood.TreeLikelihood.updateParams" title="Permalink to this definition">¶</a></dt>
<dd><p>Update model parameters and re-compute likelihoods.</p>
<p>This method is the <strong>only</strong> acceptable way to update model
parameters. The likelihood is re-computed as needed
by this method.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>newvalues</cite> (dict)</dt><dd><p>A dictionary keyed by param name and with value as new
value to set. Each parameter name must either be a
valid model parameter (in <cite>model.freeparams</cite>).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-phydmslib.utils">
<span id="phydmslib-utils-module"></span><h2>phydmslib.utils module<a class="headerlink" href="#module-phydmslib.utils" title="Permalink to this headline">¶</a></h2>
<p>Utilities for <code class="docutils literal notranslate"><span class="pre">phydmslib</span></code>.</p>
<dl class="function">
<dt id="phydmslib.utils.BenjaminiHochbergCorrection">
<code class="sig-prename descclassname">phydmslib.utils.</code><code class="sig-name descname">BenjaminiHochbergCorrection</code><span class="sig-paren">(</span><em class="sig-param">pvals</em>, <em class="sig-param">fdr</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.utils.BenjaminiHochbergCorrection" title="Permalink to this definition">¶</a></dt>
<dd><p>Benjamini-Hochberg procedure to control false discovery rate.</p>
<p>Calling arguments:</p>
<dl>
<dt><em>pvals</em><span class="classifier">a list of tuples of <em>(label, p)</em>  where <em>label</em> is some label</span></dt><dd><p>assignedto each data point, and <em>p</em> is the corresponding <em>P-value</em>.</p>
</dd>
</dl>
<p><em>fdr</em> : the desired false discovery rate</p>
<dl class="simple">
<dt>The return value is the 2-tuple <em>(pcutoff, significantlabels)</em>. After</dt><dd><p>applyingthe algorithm, all data points with <em>p &lt;= pcutoff</em> are declared
significant. The labels for these data points are in <em>significantlabels</em>.
If there are no significant sites, <em>pcutoff</em> is returned as the maximum
P-value that would have made a single point significant.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="phydmslib.utils.modelComparisonDataFrame">
<code class="sig-prename descclassname">phydmslib.utils.</code><code class="sig-name descname">modelComparisonDataFrame</code><span class="sig-paren">(</span><em class="sig-param">modelcomparisonfile</em>, <em class="sig-param">splitparams</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.utils.modelComparisonDataFrame" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts <code class="docutils literal notranslate"><span class="pre">modelcomparison.md</span></code> file to <cite>pandas</cite> DataFrame.</p>
<p>Running <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> creates a file with the suffix
<code class="docutils literal notranslate"><span class="pre">modelcomparison.md</span></code>. This function converts that file into a
DataFrame that is easy to handle for downstream analysis.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><cite>modelcomparisonfile</cite> (str)</dt><dd><p>The name of the <code class="docutils literal notranslate"><span class="pre">modelcomparison.md</span></code> file.</p>
</dd>
<dt><cite>splitparams</cite> (bool)</dt><dd><p>If <cite>True</cite>, create a new column for each model param in
the <cite>ParamValues</cite> column, with values of <cite>NaN</cite> if that
model does not have such a parameter.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>A <cite>pandas</cite> DataFrame with the information in the model
comparison file.</p>
</dd>
</dl>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span>
<span class="gp">... </span>        <span class="s1">&#39;| Model | deltaAIC | LogLikelihood | nParams | ParamValues  |&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s1">&#39;|-------|----------|---------------|---------|--------------|&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s1">&#39;| ExpCM | 0.00     | -1000.00      | 7       | x=1.0, y=2.0 |&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="s1">&#39;| YNGKP | 10.2     | -1005.10      | 7       | x=1.3, z=0.1 |&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="p">]))</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">df_split</span> <span class="o">=</span> <span class="n">modelComparisonDataFrame</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">splitparams</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">df_nosplit</span> <span class="o">=</span> <span class="n">modelComparisonDataFrame</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">splitparams</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_nosplit</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[[</span><span class="s1">&#39;ExpCM&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="s1">&#39;x=1.0, y=2.0&#39;</span><span class="p">],</span>
<span class="gp">... </span>         <span class="p">[</span><span class="s1">&#39;YNGKP&#39;</span><span class="p">,</span> <span class="mf">10.2</span><span class="p">,</span> <span class="o">-</span><span class="mf">1005.1</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="s1">&#39;x=1.3, z=0.1&#39;</span><span class="p">]],</span>
<span class="gp">... </span>        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Model&#39;</span><span class="p">,</span> <span class="s1">&#39;deltaAIC&#39;</span><span class="p">,</span> <span class="s1">&#39;LogLikelihood&#39;</span><span class="p">,</span>
<span class="gp">... </span>                 <span class="s1">&#39;nParams&#39;</span><span class="p">,</span> <span class="s1">&#39;ParamValues&#39;</span><span class="p">]))</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_split</span><span class="o">.</span><span class="n">equals</span><span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[[</span><span class="s1">&#39;ExpCM&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span><span class="p">],</span>
<span class="gp">... </span>         <span class="p">[</span><span class="s1">&#39;YNGKP&#39;</span><span class="p">,</span> <span class="mf">10.2</span><span class="p">,</span> <span class="o">-</span><span class="mf">1005.1</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mf">1.3</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">]],</span>
<span class="gp">... </span>        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Model&#39;</span><span class="p">,</span> <span class="s1">&#39;deltaAIC&#39;</span><span class="p">,</span> <span class="s1">&#39;LogLikelihood&#39;</span><span class="p">,</span>
<span class="gp">... </span>                 <span class="s1">&#39;nParams&#39;</span><span class="p">,</span> <span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="s1">&#39;z&#39;</span><span class="p">]))</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="module-phydmslib.weblogo">
<span id="phydmslib-weblogo-module"></span><h2>phydmslib.weblogo module<a class="headerlink" href="#module-phydmslib.weblogo" title="Permalink to this headline">¶</a></h2>
<div class="section" id="weblogo-module">
<h3><code class="docutils literal notranslate"><span class="pre">weblogo</span></code> module<a class="headerlink" href="#weblogo-module" title="Permalink to this headline">¶</a></h3>
<p>Module for making sequence logos with the <em>weblogolib</em> package distributed with
<code class="docutils literal notranslate"><span class="pre">weblogo</span></code>  This module interfaces with the <em>weblogolib</em> API,
and so is only known to work with <em>weblogolib</em> version 3.4 and 3.5.</p>
<p>Written by Jesse Bloom and Mike Doud</p>
<dl class="function">
<dt id="phydmslib.weblogo.ChargeColorMapping">
<code class="sig-prename descclassname">phydmslib.weblogo.</code><code class="sig-name descname">ChargeColorMapping</code><span class="sig-paren">(</span><em class="sig-param">maptype='jet'</em>, <em class="sig-param">reverse=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.weblogo.ChargeColorMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps amino-acid charge at neutral pH to colors. 
Currently does not use the keyword arguments for <em>maptype</em>
or <em>reverse</em> but accepts these arguments to be consistent
with KyteDoolittleColorMapping and MWColorMapping for now.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.weblogo.FunctionalGroupColorMapping">
<code class="sig-prename descclassname">phydmslib.weblogo.</code><code class="sig-name descname">FunctionalGroupColorMapping</code><span class="sig-paren">(</span><em class="sig-param">maptype='jet'</em>, <em class="sig-param">reverse=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.weblogo.FunctionalGroupColorMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps amino-acid functional groups to colors.
Currently does not use the keyword arguments for <em>maptype</em>
or <em>reverse</em> but accepts these arguments to be consistent
with the other mapping functions, which all get called with 
these arguments.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.weblogo.KyteDoolittleColorMapping">
<code class="sig-prename descclassname">phydmslib.weblogo.</code><code class="sig-name descname">KyteDoolittleColorMapping</code><span class="sig-paren">(</span><em class="sig-param">maptype='jet'</em>, <em class="sig-param">reverse=True</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.weblogo.KyteDoolittleColorMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps amino-acid hydrophobicities to colors.</p>
<p>Uses the Kyte-Doolittle hydrophobicity scale defined by:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">J</span><span class="o">.</span> <span class="n">Kyte</span> <span class="o">&amp;</span> <span class="n">R</span><span class="o">.</span> <span class="n">F</span><span class="o">.</span> <span class="n">Doolittle</span><span class="p">:</span> 
<span class="s2">&quot;A simple method for displaying the hydropathic character of a protein.&quot;</span> 
<span class="n">J</span> <span class="n">Mol</span> <span class="n">Biol</span><span class="p">,</span> <span class="mi">157</span><span class="p">,</span> <span class="mi">105</span><span class="o">-</span><span class="mi">132</span>
</pre></div>
</div>
<p>More positive values indicate higher hydrophobicity, while more negative values
indicate lower hydrophobicity.</p>
<p>The returned variable is the 3-tuple <em>(cmap, mapping_d, mapper)</em>:</p>
<blockquote>
<div><ul class="simple">
<li><p><em>cmap</em> is a <code class="docutils literal notranslate"><span class="pre">pylab</span></code> <em>LinearSegmentedColorMap</em> object.</p></li>
<li><p><em>mapping_d</em> is a dictionary keyed by the one-letter amino-acid
codes. The values are the colors in CSS2 format (e.g. #FF0000
for red) for that amino acid. The value for a stop codon 
(denoted by a <code class="docutils literal notranslate"><span class="pre">*</span></code> character) is black (#000000).</p></li>
<li><p><em>mapper</em> is the actual <em>pylab.cm.ScalarMappable</em> object.</p></li>
</ul>
</div></blockquote>
<p>The optional calling argument <em>maptype</em> should specify a valid <code class="docutils literal notranslate"><span class="pre">pylab</span></code> color map.</p>
<p>The optional calling argument <em>reverse</em> specifies that we set up the color
map so that the most hydrophobic residue comes first (in the Kyte-Doolittle
scale the most hydrophobic comes last as it has the largest value). This option
is <em>True</em> by default as it seems more intuitive to have charged residues red
and hydrophobic ones blue.</p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.weblogo.LogoOverlay">
<code class="sig-prename descclassname">phydmslib.weblogo.</code><code class="sig-name descname">LogoOverlay</code><span class="sig-paren">(</span><em class="sig-param">sites</em>, <em class="sig-param">overlayfile</em>, <em class="sig-param">overlay</em>, <em class="sig-param">nperline</em>, <em class="sig-param">sitewidth</em>, <em class="sig-param">rmargin</em>, <em class="sig-param">logoheight</em>, <em class="sig-param">barheight</em>, <em class="sig-param">barspacing</em>, <em class="sig-param">fix_limits={}</em>, <em class="sig-param">fixlongname=False</em>, <em class="sig-param">overlay_cmap=None</em>, <em class="sig-param">underlay=False</em>, <em class="sig-param">scalebar=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.weblogo.LogoOverlay" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes overlay for <em>LogoPlot</em>.</p>
<p>This function creates colored bars overlay bars showing up to two
properties.
The trick of this function is to create the bars the right
size so they align when they overlay the logo plot.</p>
<p>CALLING VARIABLES:</p>
<ul class="simple">
<li><p><em>sites</em> : same as the variable of this name used by <em>LogoPlot</em>.</p></li>
<li><p><em>overlayfile</em> is a string giving the name of created PDF file containing
the overlay. It must end in the extension <code class="docutils literal notranslate"><span class="pre">.pdf</span></code>.</p></li>
<li><p><em>overlay</em> : same as the variable of this name used by <em>LogoPlot</em>.</p></li>
<li><p><em>nperline</em> : same as the variable of this name used by <em>LogoPlot</em>.</p></li>
<li><p><em>sitewidth</em> is the width of each site in points.</p></li>
<li><p><em>rmargin</em> is the right margin in points.</p></li>
<li><p><em>logoheight</em> is the total height of each logo row in points.</p></li>
<li><p><em>barheight</em> is the total height of each bar in points.</p></li>
<li><p><em>barspacing</em> is the vertical spacing between bars in points.</p></li>
<li><p><em>fix_limits</em> has the same meaning of the variable of this name used by <em>LogoPlot</em>.</p></li>
<li><p><em>fixlongname</em> has the same meaning of the variable of this name used by <em>LogoPlot</em>.</p></li>
<li><p><em>overlay_cmap</em> has the same meaning of the variable of this name used by <em>LogoPlot</em>.</p></li>
<li><p><em>underlay</em> is a bool. If <cite>True</cite>, make an underlay rather than an overlay.</p></li>
<li><p><a href="#id4"><span class="problematic" id="id5">*</span></a>scalebar: if not <cite>False</cite>, is 2-tuple <cite>(scalebarheight, scalebarlabel)</cite>
where <cite>scalebarheight</cite> is in points.</p></li>
</ul>
</dd></dl>

<dl class="function">
<dt id="phydmslib.weblogo.LogoPlot">
<code class="sig-prename descclassname">phydmslib.weblogo.</code><code class="sig-name descname">LogoPlot</code><span class="sig-paren">(</span><em class="sig-param">sites</em>, <em class="sig-param">datatype</em>, <em class="sig-param">data</em>, <em class="sig-param">plotfile</em>, <em class="sig-param">nperline</em>, <em class="sig-param">numberevery=10</em>, <em class="sig-param">allowunsorted=False</em>, <em class="sig-param">ydatamax=1.01</em>, <em class="sig-param">overlay=None</em>, <em class="sig-param">fix_limits={}</em>, <em class="sig-param">fixlongname=False</em>, <em class="sig-param">overlay_cmap=None</em>, <em class="sig-param">ylimits=None</em>, <em class="sig-param">relativestackheight=1</em>, <em class="sig-param">custom_cmap='jet'</em>, <em class="sig-param">map_metric='kd'</em>, <em class="sig-param">noseparator=False</em>, <em class="sig-param">underlay=False</em>, <em class="sig-param">scalebar=False</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.weblogo.LogoPlot" title="Permalink to this definition">¶</a></dt>
<dd><p>Create sequence logo showing amino-acid or nucleotide preferences.</p>
<p>The heights of each letter is equal to the preference of
that site for that amino acid or nucleotide.</p>
<p>Note that stop codons may or may not be included in the logo
depending on whether they are present in <em>pi_d</em>.</p>
<p>CALLING VARIABLES:</p>
<ul>
<li><p><em>sites</em> is a list of all of the sites that are being included
in the logo, as strings. They must be in natural sort or an error
will be raised <strong>unless</strong> <em>allowunsorted</em> is <em>True</em>. The sites
in the plot are ordered in the same arrangement
listed in <em>sites</em>. These should be <strong>strings</strong>, not integers.</p></li>
<li><p><em>datatype</em> should be one of the following strings:</p>
<blockquote>
<div><ul class="simple">
<li><p>‘prefs’ for preferences</p></li>
<li><p>‘diffprefs’ for differential preferences</p></li>
<li><p>‘diffsel’ for differential selection</p></li>
</ul>
</div></blockquote>
</li>
<li><p><em>data</em> is a dictionary that has a key for every entry in
<em>sites</em>. For every site <em>r</em> in <em>sites</em>, <em>sites[r][x]</em>
is the value for character <em>x</em>. 
Preferences must sum to one; differential preferences to zero.
All sites must have the same set of characters. The characters
must be the set of nucleotides or amino acids with or without
stop codons.</p></li>
<li><p><em>plotfile</em> is a string giving the name of the created PDF file 
of the logo plot.
It must end in the extension <code class="docutils literal notranslate"><span class="pre">.pdf</span></code>.</p></li>
<li><p><em>nperline</em> is the number of sites per line. Often 40 to 80 are good values.</p></li>
<li><p><em>numberevery</em> is specifies how frequently we put labels for the sites on
x-axis.</p></li>
<li><p><em>allowunsorted</em> : if <em>True</em> then we allow the entries in <em>sites</em> to 
<strong>not</strong> be sorted. This means that the logo plot will <strong>not</strong> have
sites in sorted order.</p></li>
<li><p><em>ydatamax</em> : meaningful only if <em>datatype</em> is ‘diffprefs’. In this case, it gives
the maximum that the logo stacks extend in the positive and negative directions.
Cannot be smaller than the maximum extent of the differential preferences.</p></li>
<li><p><em>ylimits</em>: is <strong>mandatory</strong> if <em>datatype</em> is ‘diffsel’, and meaningless 
otherwise. It is <em>(ymin, ymax)</em> where <em>ymax &gt; 0 &gt; ymin</em>, and gives extent 
of the data in the positive and negative directions. Must encompass the 
actual maximum and minimum of the data.</p></li>
<li><p><em>overlay</em> : make overlay bars that indicate other properties for
the sites. If you set to something other than <cite>None</cite>, it should be
a list giving one to three properties. Each property is a tuple:
<em>(prop_d, shortname, longname)</em> where:</p>
<blockquote>
<div><ul>
<li><p><em>prop_d</em> is a dictionary keyed by site numbers that are in <em>sites</em>.
For each <em>r</em> in <em>sites</em>, <em>prop_d[r]</em> gives the value of the property,
or if there is no entry in <em>prop_d</em> for <em>r</em>, then the property
is undefined and is colored white. Properties can either be:</p>
<blockquote>
<div><ul class="simple">
<li><p>continuous: in this case, all of the values should be numbers.</p></li>
<li><p>discrete : in this case, all of the values should be strings.
While in practice, if you have more than a few discrete
categories (different strings), the plot will be a mess.</p></li>
</ul>
</div></blockquote>
</li>
<li><p><em>shortname</em> : short name for the property; will not format well
if more than 4 or 5 characters.</p></li>
<li><p><em>longname</em> : longer name for property used on axes label. Can be the
same as <em>shortname</em> if you don’t need a different long name.</p></li>
<li><p>In the special case where both <em>shortname</em> and <em>longname</em> are 
the string <cite>wildtype</cite>, then rather than an overlay bar we
right the one-character wildtype identity in <cite>prop_d</cite> for each
site.</p></li>
</ul>
</div></blockquote>
</li>
<li><p><em>fix_limits</em> is only meaningful if <em>overlay</em> is being used. In this case, for any
<em>shortname</em> in <em>overlay</em> that also keys an entry in <em>fix_limits</em>, we use
<em>fix_limits[shortname]</em> to set the limits for <em>shortname</em>. Specifically,
<em>fix_limits[shortname]</em> should be the 2-tuple <em>(ticks, ticknames)</em>. <em>ticks</em>
should be a list of tick locations (numbers) and <em>ticknames</em> should be a list of
the corresponding tick label for that tick.</p></li>
<li><p>If <em>fixlongname</em> is <em>True</em>, then we use the <em>longname</em> in <em>overlay</em> exactly as written;
otherwise we add a parenthesis indicating the <em>shortname</em> for which this <em>longname</em>
stands.</p></li>
<li><p><em>overlay_cmap</em> can be the name of a valid <em>matplotlib.colors.Colormap</em>, such as the
string <em>jet</em> or <em>bwr</em>. Otherwise, it can be <em>None</em> and a (hopefully) good choice will 
be made for you.</p></li>
<li><p><em>custom_cmap</em> can be the name of a valid <em>matplotlib.colors.Colormap</em> which will be
used to color amino-acid one-letter codes in the logoplot by the <em>map_metric</em> when
either ‘kd’ or ‘mw’ is used as <em>map_metric</em>. If <em>map_metric</em> is ‘singlecolor’,
then should be string giving the color to plot.</p></li>
<li><p><em>relativestackheight</em> indicates how high the letter stack is relative to
the default. The default is multiplied by this number, so make it &gt; 1
for a higher letter stack.</p></li>
<li><p><em>map_metric</em> specifies the amino-acid property metric used to map colors to amino-acid
letters. Valid options are ‘kd’ (Kyte-Doolittle hydrophobicity scale, default), ‘mw’ 
(molecular weight), ‘functionalgroup’ (functional groups: small, nucleophilic, hydrophobic,
aromatic, basic, acidic, and amide), ‘charge’ (charge at neutral pH), and
‘singlecolor’. If ‘charge’ is used, then the
argument for <em>custom_cmap</em> will no longer be meaningful, since ‘charge’ uses its own 
blue/black/red colormapping. Similarly, ‘functionalgroup’ uses its own colormapping.</p></li>
<li><p><em>noseparator</em> is only meaningful if <em>datatype</em> is ‘diffsel’ or ‘diffprefs’.
If it set to <em>True</em>, then we do <strong>not</strong> print a black horizontal line to
separate positive and negative values.</p></li>
<li><p><em>underlay</em> if <cite>True</cite> then make an underlay rather than an overlay.</p></li>
<li><p><em>scalebar</em>: show a scale bar. If <cite>False</cite>, no scale bar shown. Otherwise
should be a 2-tuple of <cite>(scalebarlen, scalebarlabel)</cite>. Currently only
works when data is <cite>diffsel</cite>.</p></li>
</ul>
</dd></dl>

<dl class="function">
<dt id="phydmslib.weblogo.MWColorMapping">
<code class="sig-prename descclassname">phydmslib.weblogo.</code><code class="sig-name descname">MWColorMapping</code><span class="sig-paren">(</span><em class="sig-param">maptype='jet'</em>, <em class="sig-param">reverse=True</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.weblogo.MWColorMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps amino-acid molecular weights to colors. Otherwise, this
function is identical to <em>KyteDoolittleColorMapping</em></p>
</dd></dl>

<dl class="function">
<dt id="phydmslib.weblogo.SingleColorMapping">
<code class="sig-prename descclassname">phydmslib.weblogo.</code><code class="sig-name descname">SingleColorMapping</code><span class="sig-paren">(</span><em class="sig-param">maptype='#999999'</em><span class="sig-paren">)</span><a class="headerlink" href="#phydmslib.weblogo.SingleColorMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps all amino acids to the single color given by <cite>maptype</cite>.</p>
</dd></dl>

</div>
</div>
<div class="section" id="module-phydmslib">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-phydmslib" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">phydmslib</span></code> is the Python API for <code class="docutils literal notranslate"><span class="pre">phydms</span></code>.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">phydms</h1>
    
  </a>
</p>



<p class="blurb">Software for phylogenetic analyses informed by deep mutational scanning.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=phydms&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/phydms">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/phydms.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/phydms.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="howtouse.html">how to use <code class="docutils literal notranslate"><span class="pre">phydms</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="phydms_prog.html">the <code class="docutils literal notranslate"><span class="pre">phydms</span></code> program</a></li>
<li class="toctree-l1"><a class="reference internal" href="auxiliary_progs.html">auxiliary programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="ExpCM.html"><strong>Exp</strong>erimentally Informed <strong>C</strong>odon <strong>M</strong>odels</a></li>
<li class="toctree-l1"><a class="reference internal" href="implementation.html">Numerical implementation</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html"><code class="docutils literal notranslate"><span class="pre">phydmslib</span></code> API</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgments and citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="indices_and_tables.html">Indices and tables</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="api.html"><code class="docutils literal notranslate"><span class="pre">phydmslib</span></code> API</a><ul>
      <li>Previous: <a href="api.html" title="previous chapter"><code class="docutils literal notranslate"><span class="pre">phydmslib</span></code> API</a></li>
      <li>Next: <a href="acknowledgments.html" title="next chapter">Acknowledgments and citations</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015-2020, Sarah Hilton and Jesse Bloom.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/phydmslib.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/phydms" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>