
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>the phydms_logoplot program &#8212; phydms 2.4.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Experimentally Informed Codon Models" href="ExpCM.html" />
    <link rel="prev" title="the phydms_prepalignment program" href="phydms_prepalignment.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="the-phydms-logoplot-program">
<span id="phydms-logoplot"></span><h1><a class="toc-backref" href="#id1">the <code class="docutils literal notranslate"><span class="pre">phydms_logoplot</span></code> program</a><a class="headerlink" href="#the-phydms-logoplot-program" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#the-phydms-logoplot-program" id="id1">the <code class="docutils literal notranslate"><span class="pre">phydms_logoplot</span></code> program</a></p>
<ul>
<li><p><a class="reference internal" href="#overview" id="id2">Overview</a></p></li>
<li><p><a class="reference internal" href="#command-line-usage" id="id3">Command-line usage</a></p>
<ul>
<li><p><a class="reference internal" href="#Positional Arguments" id="id4">Positional Arguments</a></p></li>
<li><p><a class="reference internal" href="#Named Arguments" id="id5">Named Arguments</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#output-plot" id="id6">Output plot</a></p>
<ul>
<li><p><a class="reference internal" href="#plotting-preferences" id="id7">Plotting preferences</a></p></li>
<li><p><a class="reference internal" href="#overlaying-omega-values-for-each-site" id="id8">Overlaying omega values for each site</a></p></li>
<li><p><a class="reference internal" href="#plotting-differential-preferences" id="id9">Plotting differential preferences</a></p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">phydms_logoplot</span></code> is a program that can be used to visualized amino-acid preferences or differential preferences using logo plots.
It relies on <a class="reference external" href="https://code.google.com/p/weblogo/">weblogo</a> to render the actual plots.</p>
<p>Running <code class="docutils literal notranslate"><span class="pre">phydms_logoplot</span></code> with the <code class="docutils literal notranslate"><span class="pre">--prefs</span></code> option allows you to visualize amino-acid preferences, which you can also re-scale by the stringency parameter (<span class="math notranslate nohighlight">\(\beta\)</span>) inferred by <code class="docutils literal notranslate"><span class="pre">phydms</span></code> using the <code class="docutils literal notranslate"><span class="pre">--stringency</span></code> option.</p>
<p>Running <code class="docutils literal notranslate"><span class="pre">phydms_logoplot</span></code> with the <code class="docutils literal notranslate"><span class="pre">--diffprefs</span></code> option allows you to visualize the differential preferences inferred by <code class="docutils literal notranslate"><span class="pre">phydms</span></code> with the <code class="docutils literal notranslate"><span class="pre">--diffprefsbysite</span></code> option.</p>
<p>For both types of data, you can also overlay diversifying selection (inferred by <code class="docutils literal notranslate"><span class="pre">phydms</span></code> using <code class="docutils literal notranslate"><span class="pre">--omegabysite</span></code>) via the <code class="docutils literal notranslate"><span class="pre">--omegabysite</span></code> option.</p>
<p>See below for information on <a class="reference internal" href="#command-line-usage">Command-line usage</a> for <code class="docutils literal notranslate"><span class="pre">phydms_logoplot</span></code>.</p>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id3">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p><p>Make logo plot of preferences or differential preferences. Uses weblogo (<a class="reference external" href="http://weblogo.threeplusone.com/">http://weblogo.threeplusone.com/</a>). The ‘phydms’ package is written by the Bloom lab (see <a class="reference external" href="https://github.com/jbloomlab/phydms/contributors">https://github.com/jbloomlab/phydms/contributors</a>). Version 2.4.0. Full documentation at <a class="reference external" href="http://jbloomlab.github.io/phydms">http://jbloomlab.github.io/phydms</a></p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">phydms_logoplot</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">(</span><span class="o">--</span><span class="n">prefs</span> <span class="n">PREFS</span> <span class="o">|</span> <span class="o">--</span><span class="n">diffprefs</span> <span class="n">DIFFPREFS</span><span class="p">)</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">stringency</span> <span class="n">STRINGENCY</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">nperline</span> <span class="n">NPERLINE</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">numberevery</span> <span class="n">NUMBEREVERY</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">mapmetric</span> <span class="p">{</span><span class="n">kd</span><span class="p">,</span><span class="n">mw</span><span class="p">,</span><span class="n">charge</span><span class="p">,</span><span class="n">functionalgroup</span><span class="p">}]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">colormap</span> <span class="n">COLORMAP</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">diffprefheight</span> <span class="n">DIFFPREFHEIGHT</span><span class="p">]</span>
                       <span class="p">[</span><span class="o">--</span><span class="n">omegabysite</span> <span class="n">OMEGABYSITE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">minP</span> <span class="n">MINP</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                       <span class="n">outfile</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3><a class="toc-backref" href="#id4">Positional Arguments</a><a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>outfile</kbd></dt>
<dd><p>Name of created PDF logo plot.</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments">
<h3><a class="toc-backref" href="#id5">Named Arguments</a><a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--prefs</kbd></dt>
<dd><p>File with amino-acid preferences; same format as input to ‘phydms’.</p>
</dd>
<dt><kbd>--diffprefs</kbd></dt>
<dd><p>File with differential preferences; in format output by ‘phydms’.</p>
</dd>
<dt><kbd>--stringency</kbd></dt>
<dd><p>Stringency parameter to re-scale prefs.</p>
<p>Default: 1</p>
<p>Running <code class="docutils literal notranslate"><span class="pre">phydms</span></code> with an <em>ExpCM</em> infers a stringency parameter, <span class="math notranslate nohighlight">\(\beta\)</span>. If you want your plotted preferences to be re-scaled by this stringency parameter, specify it here. A value of 1 corresponds to no re-scaling.</p>
</dd>
<dt><kbd>--nperline</kbd></dt>
<dd><p>Number of sites per line.</p>
<p>Default: 70</p>
<p>Increasing this number makes for wider and shorter plots, decreasing it makes for thinner and taller plots.</p>
</dd>
<dt><kbd>--numberevery</kbd></dt>
<dd><p>Number sites at this interval.</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--mapmetric</kbd></dt>
<dd><p>Possible choices: kd, mw, charge, functionalgroup</p>
<p>Metric used to color amino-acid letters. kd = Kyte-Doolittle hydrophobicity; mw = molecular weight; functionalgroup = divide in 7 groups; charge = charge at neutral pH.</p>
<p>Default: “functionalgroup”</p>
</dd>
<dt><kbd>--colormap</kbd></dt>
<dd><p>Name of <cite>matplotlib</cite> color map for amino acids when <cite>–mapmetric</cite> is ‘kd’ or ‘mw’.</p>
<p>Default: “jet”</p>
</dd>
<dt><kbd>--diffprefheight</kbd></dt>
<dd><p>Height of diffpref logo in each direction.</p>
<p>Default: 1.0</p>
<p>When using <code class="docutils literal notranslate"><span class="pre">--diffprefs</span></code>, the logo stacks will have this height in both the positive and negative direction. You will get an error if the height of a logo stack exceeds this value. Note that the maximum possible differential preference logo stack height is 1 if the preferences are completely different between the natural evolution and those provided in the site-specific amino-acid preferences file.</p>
</dd>
<dt><kbd>--omegabysite</kbd></dt>
<dd><p>Overlay omega on logo plot. Specify ‘<a href="#id1"><span class="problematic" id="id2">*</span></a>_omegabysite.txt’ file from ‘phydms’.</p>
</dd>
<dt><kbd>--minP</kbd></dt>
<dd><p>Min plotted P-value for ‘–omegabysite’ overlay.</p>
<p>Default: 0.0001</p>
</dd>
<dt><kbd>-v, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="output-plot">
<h2><a class="toc-backref" href="#id6">Output plot</a><a class="headerlink" href="#output-plot" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">phydms_logoplot</span></code> prints some basic information to the screen.
But its primary output is a PDF plot that visualizes the preferences or differential preferences.</p>
<div class="section" id="plotting-preferences">
<h3><a class="toc-backref" href="#id7">Plotting preferences</a><a class="headerlink" href="#plotting-preferences" title="Permalink to this headline">¶</a></h3>
<p>Using the <code class="docutils literal notranslate"><span class="pre">--prefs</span></code> option plots amino-acid preferences.
This can either be used to visualize the preferences that you are using as input to <code class="docutils literal notranslate"><span class="pre">phydms</span></code>, or to visualize how these preferences look after being re-scaled by the stringency parameter (<span class="math notranslate nohighlight">\(\beta\)</span>) inferred by <code class="docutils literal notranslate"><span class="pre">phydms</span></code>.</p>
<p>You can do this with a command like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">phydms_logoplot</span> <span class="o">--</span><span class="n">prefs</span> <span class="n">prefsfile</span><span class="o">.</span><span class="n">tsv</span> <span class="n">prefs_logoplot</span><span class="o">.</span><span class="n">pdf</span> <span class="o">--</span><span class="n">stringency</span> <span class="mf">2.99</span> <span class="o">--</span><span class="n">nperline</span> <span class="mi">72</span> <span class="o">--</span><span class="n">mapmetric</span> <span class="n">charge</span>
</pre></div>
</div>
<p>Here is the resulting plot:</p>
<a class="reference internal image-reference" href="_images/prefs_logoplot.jpg"><img alt="jpeg version of logoplot pdf." class="align-center" src="_images/prefs_logoplot.jpg" style="width: 95%;" /></a>
<p>The preferences in this plot are re-scaled by the stringency parameter of <span class="math notranslate nohighlight">\(\beta = 2.99\)</span>, which we got from the <code class="docutils literal notranslate"><span class="pre">*_modelparams.txt</span></code> file after running <code class="docutils literal notranslate"><span class="pre">phydms</span></code>.</p>
<p>Note that we have set <code class="docutils literal notranslate"><span class="pre">--mapmetric</span></code> to color amino acids by charge. They can also be colored using other schemes.</p>
</div>
<div class="section" id="overlaying-omega-values-for-each-site">
<h3><a class="toc-backref" href="#id8">Overlaying omega values for each site</a><a class="headerlink" href="#overlaying-omega-values-for-each-site" title="Permalink to this headline">¶</a></h3>
<p>One use of <code class="docutils literal notranslate"><span class="pre">phydms</span></code> is to identify diversifying selection via the <code class="docutils literal notranslate"><span class="pre">--omegabysite</span></code> values.
You may want to overlay these data onto the logoplots of either preferences or differential preferences.</p>
<p>Here we overlay the diversifying selection the same preferences logo plot shown above:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">phydms_logoplot</span> <span class="o">--</span><span class="n">prefs</span> <span class="n">prefsfile</span><span class="o">.</span><span class="n">tsv</span> <span class="n">omegaoverlay_logoplot</span><span class="o">.</span><span class="n">pdf</span> <span class="o">--</span><span class="n">stringency</span> <span class="mf">2.99</span> <span class="o">--</span><span class="n">nperline</span> <span class="mi">72</span> <span class="o">--</span><span class="n">mapmetric</span> <span class="n">kd</span> <span class="o">--</span><span class="n">omegabysite</span> <span class="n">omegabysite</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">omegabysite.txt</span></code> is the file produced by <code class="docutils literal notranslate"><span class="pre">phydms</span></code>. The created plot looks like this:</p>
<a class="reference internal image-reference" href="_images/omegaoverlay_logoplot.jpg"><img alt="jpeg version of omega overlay pdf." class="align-center" src="_images/omegaoverlay_logoplot.jpg" style="width: 95%;" /></a>
<p>Note that the plot shows the <em>P</em>-value that <span class="math notranslate nohighlight">\(\omega_r\)</span> is greater than or less than one (recall that the <em>P</em>-values are more meaningful than the actual <span class="math notranslate nohighlight">\(\omega_r\)</span> values for this kind of analysis).
These <em>P</em>-values are <strong>not</strong> corrected for multiple testing (for that, look at the <em>Q</em>-values output by <code class="docutils literal notranslate"><span class="pre">phydms</span></code>).</p>
<p>Note also that we have used a different <code class="docutils literal notranslate"><span class="pre">--mapmetric</span></code>, this time coloring amino acids by hydrophobicity on the Kyte-Doolittle scale.</p>
</div>
<div class="section" id="plotting-differential-preferences">
<h3><a class="toc-backref" href="#id9">Plotting differential preferences</a><a class="headerlink" href="#plotting-differential-preferences" title="Permalink to this headline">¶</a></h3>
<p>Using the <code class="docutils literal notranslate"><span class="pre">--diffprefs</span></code> option plots the differential amino-acid preferences that can be inferred by <code class="docutils literal notranslate"><span class="pre">phydms</span></code> using <code class="docutils literal notranslate"><span class="pre">--diffprefsbysite</span></code>.
For each site, there is a logo stack with the height of each amino-acid single-letter code proportional to the differential preference <span class="math notranslate nohighlight">\(\Delta\pi_{r,a}\)</span>.
The y-scale is determined by <code class="docutils literal notranslate"><span class="pre">--diffprefheight</span></code>; note that the maximum possible height in each direction is one if the amino-acid preferences are completely shifted.</p>
<p>Here is an example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">phydms_logoplot</span> <span class="o">--</span><span class="n">diffprefs</span> <span class="n">diffprefsbysite</span><span class="o">.</span><span class="n">txt</span> <span class="n">diffprefs_logplot</span><span class="o">.</span><span class="n">pdf</span> <span class="o">--</span><span class="n">nperline</span> <span class="mi">72</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="_images/diffprefs_logoplot.jpg"><img alt="jpeg version of diff prefs pdf." class="align-center" src="_images/diffprefs_logoplot.jpg" style="width: 95%;" /></a>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">phydms</h1>
    
  </a>
</p>



<p class="blurb">Software for phylogenetic analyses informed by deep mutational scanning.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=phydms&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/phydms">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/phydms.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/phydms.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="howtouse.html">how to use <code class="docutils literal notranslate"><span class="pre">phydms</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="phydms_prog.html">the <code class="docutils literal notranslate"><span class="pre">phydms</span></code> program</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="auxiliary_progs.html">auxiliary programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="ExpCM.html"><strong>Exp</strong>erimentally Informed <strong>C</strong>odon <strong>M</strong>odels</a></li>
<li class="toctree-l1"><a class="reference internal" href="implementation.html">Numerical implementation</a></li>
<li class="toctree-l1"><a class="reference internal" href="phydmslib.html">phydmslib package</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgments and citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="indices_and_tables.html">Indices and tables</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="auxiliary_progs.html">auxiliary programs</a><ul>
      <li>Previous: <a href="phydms_prepalignment.html" title="previous chapter">the <code class="docutils literal notranslate"><span class="pre">phydms_prepalignment</span></code> program</a></li>
      <li>Next: <a href="ExpCM.html" title="next chapter"><strong>Exp</strong>erimentally Informed <strong>C</strong>odon <strong>M</strong>odels</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015-2020, Sarah Hilton and Jesse Bloom.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/phydms_logoplot.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/phydms" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>