
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>the phydms_comprehensive program &#8212; phydms 2.3.4 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="the phydms_prepalignment program" href="phydms_prepalignment.html" />
    <link rel="prev" title="auxiliary programs" href="auxiliary_progs.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="the-phydms-comprehensive-program">
<span id="phydms-comprehensive-prog"></span><h1><a class="toc-backref" href="#id1">the <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> program</a><a class="headerlink" href="#the-phydms-comprehensive-program" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#the-phydms-comprehensive-program" id="id1">the <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> program</a></p>
<ul>
<li><p><a class="reference internal" href="#overview" id="id2">Overview</a></p></li>
<li><p><a class="reference internal" href="#command-line-usage" id="id3">Command-line usage</a></p>
<ul>
<li><p><a class="reference internal" href="#Positional Arguments" id="id4">Positional Arguments</a></p></li>
<li><p><a class="reference internal" href="#Named Arguments" id="id5">Named Arguments</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#output-files" id="id6">Output files</a></p>
<ul>
<li><p><a class="reference internal" href="#log-file" id="id7">Log file</a></p></li>
<li><p><a class="reference internal" href="#model-comparison-files" id="id8">Model comparison files</a></p></li>
<li><p><a class="reference internal" href="#phydms-output-for-each-model" id="id9"><code class="docutils literal notranslate"><span class="pre">phydms</span></code> output for each model</a></p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> is a program that simplifies usage of <code class="docutils literal notranslate"><span class="pre">phydms</span></code> for standard analyses. Essentially, <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> runs <code class="docutils literal notranslate"><span class="pre">phydms</span></code> for several different models to enable model comparisons and identify selection.</p>
<p>In its simplest usage, you simply provide <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> with an alignment and one or more files giving site-specific amino-acid preferences.
The program then first uses <code class="docutils literal notranslate"><span class="pre">RAxML</span></code> to infer a tree under the <em>GTRCAT</em> model.
Alternatively, you can specify the tree using <code class="docutils literal notranslate"><span class="pre">--tree</span></code> flag.
For each set of site-specific amino-acid preferences, <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> optimizes the tree with the following models:</p>
<blockquote>
<div><ul class="simple">
<li><p><em>ExpCM</em></p></li>
<li><p><em>ExpCM</em> with a gamma-distributed <span class="math notranslate nohighlight">\(\omega\)</span> (if using the <code class="docutils literal notranslate"><span class="pre">--gammaomega</span></code> flag).</p></li>
<li><p><em>ExpCM</em> with a gamma-distributed <span class="math notranslate nohighlight">\(\beta\)</span> (if using the <code class="docutils literal notranslate"><span class="pre">--gammabeta</span></code> flag).</p></li>
<li><p><em>YNGKP_M0</em></p></li>
<li><p><em>YNGKP_M5</em></p></li>
</ul>
</div></blockquote>
<p>It also runs each of these <em>ExpCM</em> with averaged preferences as a control.
Finally, it creates summaries that enable comparison among the models.</p>
<p>You could get all of this output by simply running <code class="docutils literal notranslate"><span class="pre">phydms</span></code> repeatedly, but using <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> automates this process for you.</p>
<p>See below for information on <a class="reference internal" href="#command-line-usage">Command-line usage</a> and <a class="reference internal" href="#output-files">Output files</a>.</p>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id3">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p><p>Comprehensive phylogenetic model comparison and detection of selection informed by deep mutational scanning data. This program runs ‘phydms’ repeatedly to compare substitution models and detect selection. The ‘phydms’ package is written by the Bloom lab (see <a class="reference external" href="https://github.com/jbloomlab/phydms/contributors">https://github.com/jbloomlab/phydms/contributors</a>). Version 2.3.3. Full documentation at <a class="reference external" href="http://jbloomlab.github.io/phydms">http://jbloomlab.github.io/phydms</a></p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">phydms_comprehensive</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">(</span><span class="o">--</span><span class="n">raxml</span> <span class="n">RAXML</span> <span class="o">|</span> <span class="o">--</span><span class="n">tree</span> <span class="n">TREE</span><span class="p">)</span> <span class="p">[</span><span class="o">--</span><span class="n">ncpus</span> <span class="n">NCPUS</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">brlen</span> <span class="p">{</span><span class="n">scale</span><span class="p">,</span><span class="n">optimize</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">omegabysite</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">diffprefsbysite</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gammaomega</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gammabeta</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">avgprefs</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">randprefs</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                            <span class="n">outprefix</span> <span class="n">alignment</span> <span class="n">prefsfiles</span> <span class="p">[</span><span class="n">prefsfiles</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3><a class="toc-backref" href="#id4">Positional Arguments</a><a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>outprefix</kbd></dt>
<dd><p>Output file prefix.</p>
<p>This prefix can be a directory name (e.g. <code class="docutils literal notranslate"><span class="pre">my_directory/</span></code>) if you want to create a new directory. See <a class="reference internal" href="#output-files">Output files</a> for a description of the created files.</p>
</dd>
<dt><kbd>alignment</kbd></dt>
<dd><p>Existing FASTA file with aligned codon sequences.</p>
<p>The alignment must meet the same specifications described in the <code class="docutils literal notranslate"><span class="pre">phydms</span></code> documentation for the argument of the same name (see <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>).</p>
</dd>
<dt><kbd>prefsfiles</kbd></dt>
<dd><p>Existing files with site-specific amino-acid preferences.</p>
<p>Provide the name of one or more files giving site-specific amino-acid preferences. These files should meet the same specifications described in the <code class="docutils literal notranslate"><span class="pre">phydms</span></code> documentation for the <em>prefsfile</em> that should accompany an <em>ExpCM</em> (see <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>).</p>
</dd>
</dl>
</div>
<div class="section" id="Named Arguments">
<h3><a class="toc-backref" href="#id5">Named Arguments</a><a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option-list">
<dt><kbd>--raxml</kbd></dt>
<dd><p>Path to RAxML (e.g., ‘raxml’)</p>
<p>By default, <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> uses <code class="docutils literal notranslate"><span class="pre">RAxML</span></code> to infer a tree topology under the <em>Jukes Cantor</em> model and the command <code class="docutils literal notranslate"><span class="pre">raxml</span></code>. Use this argument to specify a path to <code class="docutils literal notranslate"><span class="pre">RAxML</span></code> other than <code class="docutils literal notranslate"><span class="pre">raxml</span></code>.  The tree inferred by <code class="docutils literal notranslate"><span class="pre">RAxML</span></code> can be found in the same directory as the other output files.</p>
</dd>
<dt><kbd>--tree</kbd></dt>
<dd><p>Existing Newick file giving input tree.</p>
<p>If you want to instead fix the tree to some existing topology, use this argument and provide the name of a file giving a valid tree in Newick format. You cannot specify both <code class="docutils literal notranslate"><span class="pre">--tree</span></code> and <code class="docutils literal notranslate"><span class="pre">--raxml</span></code>.</p>
</dd>
<dt><kbd>--ncpus</kbd></dt>
<dd><p>Use this many CPUs; -1 means all available.</p>
<p>Default: -1</p>
</dd>
<dt><kbd>--brlen</kbd></dt>
<dd><p>Possible choices: scale, optimize</p>
<p>How to handle branch lengths: scale by single parameter or optimize each one</p>
<p>Default: “optimize”</p>
</dd>
<dt><kbd>--omegabysite</kbd></dt>
<dd><p>Fit omega (dN/dS) for each site.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--diffprefsbysite</kbd></dt>
<dd><p>Fit differential preferences for each site.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--gammaomega</kbd></dt>
<dd><p>Fit ExpCM with gamma distributed omega.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--gammabeta</kbd></dt>
<dd><p>Fit ExpCM with gamma distributed beta.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--no-avgprefs</kbd></dt>
<dd><p>No fitting of models with preferences averaged across sites for ExpCM.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--randprefs</kbd></dt>
<dd><p>Include ExpCM models with randomized preferences.</p>
<p>Default: False</p>
</dd>
<dt><kbd>-v, --version</kbd></dt>
<dd><p>show program’s version number and exit</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id6">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>Running <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code> will create the following output files, all with the prefix specified by <code class="docutils literal notranslate"><span class="pre">outprefix</span></code>.</p>
<div class="section" id="log-file">
<h3><a class="toc-backref" href="#id7">Log file</a><a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>A file with the suffix <code class="docutils literal notranslate"><span class="pre">.log</span></code> will be created that summarizes the overall progress of <code class="docutils literal notranslate"><span class="pre">phydms_comprehensive</span></code>. If <code class="docutils literal notranslate"><span class="pre">outprefix</span></code> is just a directory name, this file will be called <code class="docutils literal notranslate"><span class="pre">log.log</span></code>.</p>
</div>
<div class="section" id="model-comparison-files">
<h3><a class="toc-backref" href="#id8">Model comparison files</a><a class="headerlink" href="#model-comparison-files" title="Permalink to this headline">¶</a></h3>
<p>A file with the suffix <code class="docutils literal notranslate"><span class="pre">modelcomparison.md</span></code> will be created that summarizes the model comparison.
For each model, it reports the <span class="math notranslate nohighlight">\(\Delta\rm{AIC}\)</span>, the optimized log likelihood, and the values of key parameters.
This file is in Markdown format:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">|</span> <span class="n">Model</span>                   <span class="o">|</span> <span class="n">deltaAIC</span> <span class="o">|</span> <span class="n">LogLikelihood</span> <span class="o">|</span> <span class="n">nParams</span> <span class="o">|</span> <span class="n">ParamValues</span>                                   <span class="o">|</span>
<span class="o">|-------------------------|----------|---------------|---------|-----------------------------------------------|</span>
<span class="o">|</span> <span class="n">ExpCM_NP_prefs</span>          <span class="o">|</span> <span class="mf">0.00</span>     <span class="o">|</span> <span class="o">-</span><span class="mf">3389.38</span>      <span class="o">|</span> <span class="mi">6</span>       <span class="o">|</span> <span class="n">beta</span><span class="o">=</span><span class="mf">2.99</span><span class="p">,</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">6.31</span><span class="p">,</span> <span class="n">omega</span><span class="o">=</span><span class="mf">0.78</span>             <span class="o">|</span>
<span class="o">|</span> <span class="n">averaged_ExpCM_NP_prefs</span> <span class="o">|</span> <span class="mf">2586.44</span>  <span class="o">|</span> <span class="o">-</span><span class="mf">4682.60</span>      <span class="o">|</span> <span class="mi">6</span>       <span class="o">|</span> <span class="n">beta</span><span class="o">=</span><span class="mf">0.28</span><span class="p">,</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">6.51</span><span class="p">,</span> <span class="n">omega</span><span class="o">=</span><span class="mf">0.12</span>             <span class="o">|</span>
<span class="o">|</span> <span class="n">YNGKP_M5</span>                <span class="o">|</span> <span class="mf">2599.70</span>  <span class="o">|</span> <span class="o">-</span><span class="mf">4683.23</span>      <span class="o">|</span> <span class="mi">12</span>      <span class="o">|</span> <span class="n">alpha_omega</span><span class="o">=</span><span class="mf">0.30</span><span class="p">,</span> <span class="n">beta_omega</span><span class="o">=</span><span class="mf">2.41</span><span class="p">,</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">5.84</span> <span class="o">|</span>
<span class="o">|</span> <span class="n">YNGKP_M0</span>                <span class="o">|</span> <span class="mf">2679.50</span>  <span class="o">|</span> <span class="o">-</span><span class="mf">4724.13</span>      <span class="o">|</span> <span class="mi">11</span>      <span class="o">|</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">5.79</span><span class="p">,</span> <span class="n">omega</span><span class="o">=</span><span class="mf">0.11</span>                        <span class="o">|</span>
</pre></div>
</div>
</div>
<div class="section" id="phydms-output-for-each-model">
<h3><a class="toc-backref" href="#id9"><code class="docutils literal notranslate"><span class="pre">phydms</span></code> output for each model</a><a class="headerlink" href="#phydms-output-for-each-model" title="Permalink to this headline">¶</a></h3>
<p>For each individual model, there will also be all of the expected <code class="docutils literal notranslate"><span class="pre">phydms</span></code> output files as described in <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>. These files will begin with the prefix specified by <code class="docutils literal notranslate"><span class="pre">outprefix</span></code>, which will be followed by the name of the model.</p>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">phydms</h1>
    
  </a>
</p>



<p class="blurb">Software for phylogenetic analyses informed by deep mutational scanning.</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=phydms&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/phydms">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/phydms.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/phydms.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="howtouse.html">how to use <code class="docutils literal notranslate"><span class="pre">phydms</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="phydms_prog.html">the <code class="docutils literal notranslate"><span class="pre">phydms</span></code> program</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="auxiliary_progs.html">auxiliary programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="ExpCM.html"><strong>Exp</strong>erimentally Informed <strong>C</strong>odon <strong>M</strong>odels</a></li>
<li class="toctree-l1"><a class="reference internal" href="implementation.html">Numerical implementation</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html"><code class="docutils literal notranslate"><span class="pre">phydmslib</span></code> API</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgments and citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="indices_and_tables.html">Indices and tables</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="auxiliary_progs.html">auxiliary programs</a><ul>
      <li>Previous: <a href="auxiliary_progs.html" title="previous chapter">auxiliary programs</a></li>
      <li>Next: <a href="phydms_prepalignment.html" title="next chapter">the <code class="docutils literal notranslate"><span class="pre">phydms_prepalignment</span></code> program</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2015-2019, Sarah Hilton and Jesse Bloom.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/phydms_comprehensive_prog.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/phydms" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>